<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • jamses</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">jamses</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.69.900</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="http://github.com/jmw86069/jamses/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="http://github.com/jmw86069/jamses/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.69.900" id="jamses-0069900">jamses 0.0.69.900<a class="anchor" aria-label="anchor" href="#jamses-0069900"></a></h2>
<ul><li>Bumped version for some jam packages in Imports: jamba, colorjam, multienrichjam</li>
<li>Bumped version for some jam packages in Suggests: genejam, venndir</li>
</ul><div class="section level3">
<h3 id="bug-fixes-0-0-69-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-69-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was not properly defining <code>rowData_se</code> causing <code>rowData_colnames</code> to be ignored. The bug was that although <code>"rowRanges"</code> is defined as a slot name, <code>"rowData"</code> is not. In either case the correct approach is to use accessor functions <code>rowRanges</code> and <code>rowData()</code> which handle whatever slot names are in use for the given object class.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-69-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-69-900"></a></h3>
<ul><li>
<code><a href="../reference/se_to_rowcoldata.html">se_to_rowcoldata()</a></code> - captures logic to define <code>rowData_se</code> and <code>colData_se</code> from a number of possible input classes, removing this logic from <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code>.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-69-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-69-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>now calls <code><a href="../reference/se_to_rowcoldata.html">se_to_rowcoldata()</a></code> to define <code>rowData_se</code> and <code>colData_se</code>.</li>
<li>Examples now include <code>rowData_colnames</code> to confirm proper output.</li>
<li>Now accepts <code>matrix</code> and <code>data.frame</code> input, by converting to a simple <code>SummarizedExperiment</code> object.</li>
<li>When data contains only one column it defines <code>centerby_colnames=FALSE</code> to skip the data centering, otherwise all data becomes zero.</li>
</ul></li>
<li>
<p><code><a href="../reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>Improved help docs.</li>
<li>New argument <code>hits_use_lfc=FALSE</code> which can optionally use log2 fold change values instead of directional values <code>c(-1, 0, 1)</code>. When enabled, the actual log2 fold change is multiplied by the absolute value of the hit column, resulting in the log2 fold change only for entries that met other statistical criteria.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.68.900" id="jamses-0068900">jamses 0.0.68.900<a class="anchor" aria-label="anchor" href="#jamses-0068900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-68-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-68-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Added preliminary support for <code>SingleCellExperiment</code> and <code>Seurat</code> objects. Initial support recognizes the data exactly as provided. Future options may involve filtering, transformation, etc.</li>
<li>Input now uses <code>values(rowRanges())</code> when there is no slot name <code>"rowData"</code>.</li>
<li>If either <code>rowData()</code> or <code>colData()</code> are empty (0 columns), it defines the rownames/colnames in “row”/“column”, respectively.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.67.900" id="jamses-0067900">jamses 0.0.67.900<a class="anchor" aria-label="anchor" href="#jamses-0067900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-67-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-67-900"></a></h3>
<ul><li>
<p><code><a href="../reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>
<code>rowData_colnames</code> now ignores any columns already present in the data, to prevent erroneously returning duplicated columns. This change fixes occassional error “duplicated colNames” when saving to xlsx format, caused when using <code>rowData_colnames</code>.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.66.900" id="jamses-0066900">jamses 0.0.66.900<a class="anchor" aria-label="anchor" href="#jamses-0066900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-66-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-66-900"></a></h3>
<ul><li>
<p><code><a href="../reference/shrinkDataFrame.html">shrinkDataFrame()</a></code></p>
<ul><li>Preferred over <code><a href="../reference/shrink_df.html">shrink_df()</a></code> which will be retired or converted to a wrapper to <code><a href="../reference/shrinkDataFrame.html">shrinkDataFrame()</a></code>.</li>
<li>Calls internal <code><a href="../reference/shrink_matrix.html">shrink_matrix()</a></code> to avoid <code>splicejam</code> as a dependency.</li>
<li>Tests have been added to cover the variations.</li>
</ul></li>
<li>
<p><code><a href="../reference/shrink_df.html">shrink_df()</a></code></p>
<ul><li>By default it calls <code><a href="../reference/shrinkDataFrame.html">shrinkDataFrame()</a></code> unless <code>use_new_method=TRUE</code>. (It is remarkably slower.) It may be updated over time to use more efficient and modern R methods.</li>
</ul></li>
<li><p><code><a href="../reference/list2im_opt.html">list2im_opt()</a></code> default changed to <code>do_sparse=FALSE</code></p></li>
<li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Removed <code><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">require(SummarizedExperiment)</a></code>, added package prefixes.</li>
<li>Added examples with <code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code>, <code>apply_hm_column_title=TRUE</code> and <code>hm_title_buffer=3</code>.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-66-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-66-900"></a></h3>
<ul><li>
<code><a href="../reference/shrink_matrix.html">shrink_matrix()</a></code> - copied from <code><a href="https://rdrr.io/pkg/splicejam/man/shrinkMatrix.html" class="external-link">splicejam::shrinkMatrix()</a></code> with improved default numeric function.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.65.900" id="jamses-0065900">jamses 0.0.65.900<a class="anchor" aria-label="anchor" href="#jamses-0065900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-65-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-65-900"></a></h3>
<ul><li>
<p><code><a href="../reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>Fixed bug causing mismatch in <code>sheetName</code> when only a subset of <code>sheetName</code> values were too long for the permitted Excel worksheet name.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-65-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-65-900"></a></h3>
<ul><li>
<p><code><a href="../reference/shortest_unique_abbreviation.html">shortest_unique_abbreviation()</a></code></p>
<ul><li>New function intended to provide shortest unique abbreviation for a <code>character</code> vector.</li>
<li>A neat feature enabled by default is that it will retain trailing numeric values without splitting the number, if the number is required to make the abbreviated vector unique.</li>
<li>This function can be enabled within other functions: <code>contrast2comps(..., abbreviate=TRUE)</code> and <code>save_sestats(..., abbreviate=TRUE)</code>. It is not enabled by default in either of these functions.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-65-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-65-900"></a></h3>
<ul><li>
<p><code>contrast2comps()</code>, <code><a href="../reference/contrasts_to_factors.html">contrasts_to_factors()</a></code>, and <code><a href="../reference/save_sestats.html">save_sestats()</a></code> new argument <code>abbreviate=FALSE</code>:</p>
<ul><li>Optionally calls <code><a href="../reference/shortest_unique_abbreviation.html">shortest_unique_abbreviation()</a></code> to abbreviate factor levels within each contrast.</li>
<li>It will attempt to create unique abbreviations for each factor independently.</li>
<li>If any abbreviation is not unique, or if any factor level across multiple columns is not unique, it will create global abbreviations so that each abbreviated value represents a defined factor level.</li>
<li>It should also properly handle non-standard contrasts, with each factor in order, using the appropriate abbreviation in each column.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.64.900" id="jamses-0064900">jamses 0.0.64.900<a class="anchor" aria-label="anchor" href="#jamses-0064900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-64-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-64-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>Apparently some vicious RMarkdown corner case causes <code>options("warn"=2)</code> which causes all warnings to become fatal errors. As a result, harmless <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub()</a></code> used when <code>hm_body_base</code> is auto-detected - resuling in multiple potential values. The first value is used, but presence of multiple values causes <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub()</a></code> to issue a warning, causing an error. It needed to be fixed anyway, and now it is. But <code>warn=2</code> is evil, haha.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.63.900" id="jamses-0063900">jamses 0.0.63.900<a class="anchor" aria-label="anchor" href="#jamses-0063900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-63-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-63-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> was not maintaining correct order of data matrix for a specific scenario, when <code>handle_na</code> was not <code>"none"</code>, and when the order of groups in the <code>design</code> matrix was non-standard, that is with the first group not in the first column and first row. It obviously should not matter the order of groups, but this scenario limits when the bug occurred, which is somewhat comforting.</p>
<ul><li>The cause was <code><a href="../reference/handle_na_values.html">handle_na_values()</a></code> which converted the design matrix to <code>list</code> in order to associated colnames to groups. The <code>list</code> was converted to vector, sometimes changing the order of columns returned.</li>
<li>The bug is fixed in <code><a href="../reference/handle_na_values.html">handle_na_values()</a></code> by returning columns in the same order they were provided at input.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.62.900" id="jamses-0062900">jamses 0.0.62.900<a class="anchor" aria-label="anchor" href="#jamses-0062900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-62-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-62-900"></a></h3>
<ul><li>
<p><code><a href="../reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>New argument <code>data_content=c("data", "hits")</code> changes default behavior to include a <code>"hit sheet"</code> in the output worksheets. This hit sheet includes all the “hit” columns from each contrast in one large table. For previous behavior use <code>data_content="data"</code>.</li>
<li>New argument value <code>type="list"</code> will not save to file, instead returns a <code>list</code> of <code>data.frames</code> named by worksheet name. Default is the same as before <code>type="xlsx"</code> will export to file in Excel xlsx format.</li>
<li>New arguments <code>hitRule</code>, <code>hitFormat</code>, <code>freezePaneColumn</code> merely provide some previously internal defaults, for optional customization.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.61.900" id="jamses-0061900">jamses 0.0.61.900<a class="anchor" aria-label="anchor" href="#jamses-0061900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-61-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-61-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>
<code>row_split</code> now accepts <code>data.frame</code> input</li>
<li>
<code>row_split</code> is more forgiving with vector input, now accepts a named numeric vector.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.60.900" id="jamses-0060900">jamses 0.0.60.900<a class="anchor" aria-label="anchor" href="#jamses-0060900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-60-900">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-60-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>When <code>normgroup</code> and <code>block</code> were both defined, but one <code>normgroup</code> did not have multiple values for <code>block</code>, it threw an error. In this case, when there is only one unique value for <code>block</code> within a <code>normgroup</code> it is set to <code>NULL</code> for that <code>normgroup</code> so it is not use in <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code>.</li>
<li>
<code>handle_na="full1"</code> was not performing as intended, caused when migrating away from <code>matrixStats::rowMins(x)</code> to <code>apply(x, 1, min)</code>, the logic was no longer being applied as described.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.59.900" id="jamses-0059900">jamses 0.0.59.900<a class="anchor" aria-label="anchor" href="#jamses-0059900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-59-900">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-59-900"></a></h3>
<ul><li>
<code><a href="../reference/se_detected_rows.html">se_detected_rows()</a></code> was not properly applying <code>detect_minpct</code> for the minimum percentage of replicates per group, impacting cases where groups each differed in size. Bug has been resolved.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-59-900">Changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-59-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>
<p>Arguments <code>block</code> and <code>normgroup</code> both allow three types of input, which were modified slightly to reduce the risk of values being out of sync.</p>
<ol style="list-style-type: decimal"><li>Unnamed <code>character</code> vector, must be equal to <code>length(isamples)</code>, or when <code>isamples</code> is not provided, <code>ncol(se)</code>. In this case, <code>isamples</code> or <code>colnames(se)</code> are assigned as names. <strong>This option prints an alter message</strong> suggesting that values be provided with names. However, values are named by <code>isamples</code> before any other subsetting is performed relative to <code>isamples</code>.</li>
<li>Named <code>character</code> vector, whose names match <code>colnames(se)</code> and <code>isamples</code>.</li>
<li>
<code>character</code> vector matching <code>colnames(colData(se))</code> to use one or more annotation columns, which also automatically uses <code>isamples</code> as appropriate, to maintain synchrony.</li>
</ol></li>
<li><p>Note that when using <code>normgroup</code> and/or <code>block</code> the intermediate processing can be reviewed by including <code>verbose=TRUE</code>. The output will also print each <code>correlation</code> value for each <code>normgroup</code>, when <code>block</code> is also provided.</p></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.58.900" id="jamses-0058900">jamses 0.0.58.900<a class="anchor" aria-label="anchor" href="#jamses-0058900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-58-900">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-58-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>Caused an error when using <code>normgroup</code> and <code>block</code> together, caused by the <code>block</code> not being subset to match <code>isamples</code>. Fortunately the bug would always cause failure, since it did not return an incorrect statistical result.</li>
<li>
<code>block</code> is expected to equal <code>length(isamples)</code>, and will be named as such so it can also be subset within each <code>normgroup</code>.</li>
<li>when <code>names(block)</code> are defined, they are validated to match <code>isamples</code>, and will be subset and re-ordered to match <code>isamples</code> as needed.</li>
<li>when <code>names(block)</code> is not defined, its length must equal <code>length(isamples)</code>, and then <code>names(block)</code> will be defined using <code>isamples</code>.</li>
<li>in each <code>normgroup</code> the <code>block</code> is only applied when there is more than one unique value for the relevant <code>block[isamples]</code>.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="change-to-existing-functions-0-0-58-900">change to existing functions<a class="anchor" aria-label="anchor" href="#change-to-existing-functions-0-0-58-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>Arguments <code>normgroup</code> and <code>block</code> both accept <code>colnames(colData(se))</code>, or a vector of values; however when supplying a vector of values, they must now be named using <code>isamples</code> - it no longer assumes they are the same order as <code>isamples</code>. The <code>names(block)</code> are used to confirm absolute synchrony.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.56.900" id="jamses-0056900">jamses 0.0.56.900<a class="anchor" aria-label="anchor" href="#jamses-0056900"></a></h2>
<ul><li>Added <code>edgeR</code> to Suggests, in order to enable new normalization methods <code>"TMM"</code>, <code>"TMMwsp"</code>, and <code>"RLE"</code>.</li>
</ul><div class="section level3">
<h3 id="bug-fixes-0-0-56-900">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-56-900"></a></h3>
<ul><li>
<code><a href="../reference/se_detected_rows.html">se_detected_rows()</a></code> - fixed missing package prefix with <code><a href="https://rdrr.io/pkg/jamba/man/tcount.html" class="external-link">jamba::tcount()</a></code>.</li>
<li>Added more package prefixes, most of which are used only with <code>verbose=TRUE</code>.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-56-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-56-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_normalize.html">se_normalize()</a></code>, <code><a href="../reference/matrix_normalize.html">matrix_normalize()</a></code> gained new normalization methods:</p>
<ul><li>
<p><code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code> for methods: <code>"TMM"</code>, <code>"TMMwsp"</code>, <code>"RLE"</code></p>
<ul><li>the methods perform very similarly to <code><a href="https://rdrr.io/pkg/jamma/man/jammanorm.html" class="external-link">jamma::jammanorm()</a></code>, log-ratio normalization, with some subtle differences in how they filter the features to be used during normalization. TMM,TMMwsp perform nearly identical to each other. RLE also performs more closely to both TMM than to jammanorm.</li>
<li>Ultimately this effort was not time well spent, the new methods appear to work well when data is high quality, but appear to produce questionable output when data is not consistent across samples. Meanwhile, <code>"jammanorm"</code> appears to work consistently, at least producing output that can be reasonably explained.</li>
<li>Data should be reviewed with <code><a href="https://rdrr.io/pkg/jamma/man/jammaplot.html" class="external-link">jamma::jammaplot()</a></code> MA-plots to visualize whether assumptions are met and are valid. For example, when the distribution of signal is inconsistent across samples, it appears as non-horizontal signal on MA-plots, and ultimately either the signal should be filtered prior to downstream statistics (so that the horizontal range of signal is analyzed) or another normalization method should be used in order to adjust the signal to horizontal, for example <code>"quantile"</code> normalization is intended for such purposes. In future, “LOESS” normalization may also fit this purpose.</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="../reference/se_normalize.html">se_normalize()</a></code></p>
<ul><li>New argument <code>output_method_prefix</code> to customize the method name used to formulate the output <code>assay_name</code>.</li>
<li>New argument <code>output_assay_names</code> to customize the full <code>assay_name</code> used for each output. When supplied, it overrides New argument.</li>
<li>
<code>mcols(assays(se))</code> is now populated with a <code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">S4Vectors::DataFrame</a></code> containing the <code>normalization_method</code>, <code>source_assay_name</code>, and associated <code>params</code> converted to column values. Currently experimental, but enabled by default to test the utility.</li>
<li>Converted <code><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop()</a></code> to <code><a href="https://cli.r-lib.org/reference/cli_abort.html" class="external-link">cli::cli_abort()</a></code> for improved message handling.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>New argument <code>group_line_requires_label=TRUE</code> will hide the group line when the associated label contains only whitespace or is <code>NA</code>. This defines new behavior, although for a rare scenario where there exists an empty (whitespace) group label.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.55.900" id="jamses-0055900">jamses 0.0.55.900<a class="anchor" aria-label="anchor" href="#jamses-0055900"></a></h2>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>
<code>floor_min</code> is now appropriately applied before <code>handle_na</code> specifically for the case where values may be filtered to <code>NA</code> then handled by <code>handle_na</code> using further logic.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>new argument <code>use_gridtext=TRUE</code> which by default enables markdown and limited HTML output, such as superscripts and subscripts. Implemented to render 37 degrees C using superscript <code>o</code>.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>subtle change to limit <code>rowData_colnames</code> to those values present in <code>colnames(rowData(se))</code>.</li>
<li>new argument <code>hm_name</code> which defaults to the same legend title, derived from <code>data_type</code>. However, the heatmap <code>name</code> appears in grid graphical components <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/list_components.html" class="external-link">ComplexHeatmap::list_components()</a></code> and sometimes needs to be specified in order for multiple heatmaps to have unique grid names.</li>
<li>new argument <code>hm_title</code> to provide a completely custom heatmap title. This argument is useful especially when using <code>apply_hm_column_title=TRUE</code>.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.54.900" id="jamses-0054900">jamses 0.0.54.900<a class="anchor" aria-label="anchor" href="#jamses-0054900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-54-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-54-900"></a></h3>
<ul><li>
<p><code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>when <code>group_buffer</code> is adjusted, it also adjusts placement of replicate labels <code>"n=3"</code> proportionally inside the group box.</li>
<li>new argument <code>arrow_ex</code> where <code>NULL</code> is adjusted based upon the number of contrast bumps per axis, so more parallel contrasts would cause proportionally smaller arrow sizes.</li>
<li>New argument <code>contrast_depths</code> to filter contrasts based upon depth: 1=oneway, 2=twoway. More convenient than filtering contrasts upfront.</li>
</ul></li>
<li>
<p><code><a href="../reference/contrasts_to_venn_setlists.html">contrasts_to_venn_setlists()</a></code></p>
<ul><li>Names are now user-friendly to indicate the contents of each Venn set list.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.53.900" id="jamses-0053900">jamses 0.0.53.900<a class="anchor" aria-label="anchor" href="#jamses-0053900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-53-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-53-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>new argument: <code>column_title</code> mainly used for convenience alongside the next new argument:</li>
<li>new argument: <code>apply_hm_column_title=FALSE</code> will optionally hide any existing <code>column_title</code> and instead applies the heatmap title <code>hm_title</code> to the <code>column_title</code>. This option is convenient when adding two heatmaps together, since it ensures each heatmap has its own individual title. The downside is that the usual <code>column_title</code> defined by <code>column_split</code> will not be shown.</li>
<li>new argument: <code>hm_title_buffer</code> adds an optional number of whitespace lines to the end of the heatmap title <code>hm_title</code>. This whitespace is useful when calling <code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code> which draws additional text above the heatmap.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>arguments <code>hm_title_base</code>,<code>hm_body_base</code> are now auto-detected using <code><a href="../reference/detect_heatmap_components.html">detect_heatmap_components()</a></code> by default.</li>
<li>when there is no <code>global_column_title</code> in the drawn heatmap, it means the heatmap itself did not have <code>column_title</code>, and therefore the output of this function is shifted down so it appears on the bottom of the <code>column_title</code> elements. This behavior is intended when using the new <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> argument <code>apply_hm_column_title=TRUE</code> to draw the heatmap title in the <code>column_title</code> region. It is recommended to use <code>hm_title_buffer=4</code> to add sufficient whitespace below the column title to allow room for the column group labels. In principle, the column group labels should behave the same as before when there is a <code>hm_title</code> included above the heatmap. And if both <code>global_column_title</code> and <code>column_title</code> elements are present, by default the labels from this function are placed above the <code>column_title</code> and below the <code>global_column_title</code>. <strong>whew</strong>
</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.52.900" id="jamses-0052900">jamses 0.0.52.900<a class="anchor" aria-label="anchor" href="#jamses-0052900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-52-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-52-900"></a></h3>
<ul><li>
<p><code><a href="../reference/contrasts_to_venn_setlists.html">contrasts_to_venn_setlists()</a></code></p>
<ul><li>fixed missing contrasts in certain designs</li>
<li>added method to split “equivalent” subset of contrasts using <code>max_venn_size</code> which simply subdivides remaining contrasts into subsets when needed.</li>
<li>Removed splitting contrasts solely by depth, because it produced too much nonsense.</li>
<li>Added example showing how to plot the resulting contrasts with <code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code>, which might become a useful “you are here” type of schematic to accompany a Venn diagram. Ideally, the <code>contrast_colors</code> should match the <code>set_colors</code> in <code>venndir()</code> for consistency.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-52-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-52-900"></a></h3>
<ul><li>
<code><a href="../reference/sub_split_vector.html">sub_split_vector()</a></code> is an internal function, used to split venn sets by <code>max_venn_size</code>, but in its own function because it may get some tweaks.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.51.900" id="jamses-0051900">jamses 0.0.51.900<a class="anchor" aria-label="anchor" href="#jamses-0051900"></a></h2>
<ul><li>added <code>cli</code> to package dependencies, to use messaging.</li>
</ul><div class="section level3">
<h3 id="new-functions-0-0-51-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-51-900"></a></h3>
<ul><li>
<p><code><a href="../reference/contrasts_to_factors.html">contrasts_to_factors()</a></code></p>
<ul><li>simple function to convert contrasts to a <code>data.frame</code> where each column is an experimental factor, and in the case of contrasts, at least one column per row should represent a comparison of two factor levels.</li>
</ul></li>
<li>
<p><code><a href="../reference/contrasts_to_venn_setlists.html">contrasts_to_venn_setlists()</a></code></p>
<ul><li>Convenience function to subset an overall set of contrasts to something meaningful to use in Venn diagrams. Mainly intended for RMarkdown use where the set of contrasts is not always known upfront.</li>
<li>Fixed early issue where certain sets of Venn results were not created.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-51-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-51-900"></a></h3>
<ul><li>
<p><code>contrast_names&lt;-</code></p>
<ul><li>now catches errors, printing an informative message, and returns the original data without subsetting.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.50.900" id="jamses-0050900">jamses 0.0.50.900<a class="anchor" aria-label="anchor" href="#jamses-0050900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-50-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-50-900"></a></h3>
<ul><li>
<p><code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>Fixed error when plotting a twoway contrast and one of the oneway contrast components in the opposite orientation of the twoway contrast components. Only observed in unique cases with a subset of contrasts.</li>
<li>Fixed errors ultimately caused by having duplicated <code>contrast_names</code> in the <code>SEDesign</code> object. Updated <code>contrast_names&lt;-</code> method to throw an error with clear message.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-50-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-50-900"></a></h3>
<ul><li>
<p><code><a href="../reference/filter_contrast_names.html">filter_contrast_names()</a></code></p>
<ul><li>Experimental function intended to help reduce “all-vs-all” style contrasts to those with one specific control factor level, per design factor.</li>
<li>Driven by <code>slicejam</code> design with 5 treatments, 3 genotypes, which produced 75 total oneway and twoway contrasts. Yike. Instead this function produces 30 total contrasts, only 22 oneway contrasts versus the baseline level per factor.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.49.900" id="jamses-0049900">jamses 0.0.49.900<a class="anchor" aria-label="anchor" href="#jamses-0049900"></a></h2>
<div class="section level3">
<h3 id="new-function-0-0-49-900">new function<a class="anchor" aria-label="anchor" href="#new-function-0-0-49-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_rbind.html">se_rbind()</a></code></p>
<ul><li><p>Wrapper function to automate steps required for <code>SummarizedExperiment::rbind()</code></p></li>
<li><p>Confirms that no <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code> are shared across <code>se_list</code>.</p></li>
<li><p>Optionally edits <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code> for each object, so they can be compared.</p></li>
<li><p>Determines shared <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code>, discarding other columns.</p></li>
<li><p>Optionally subsets <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code> by <code>colnames_keep</code>.</p></li>
<li>
<p>Adjusts <code>colData()</code> in one of two ways:</p>
<ol style="list-style-type: decimal"><li>
<code>colData_action="identical"</code>: keeps only colData columns which have identical values across <code>se_list</code>, discarding other columns. This option is default, since it retains common annotations.</li>
<li>
<code>colData_action="all"</code>: keeps all colData columns, using a delimiter between multiple values whenever the values differ across <code>se_list</code>. This option is useful for detailed record-keeping, but the annotations are not likely to be useful otherwise.</li>
</ol></li>
<li><p>finally it calls <code>SummarizedExperiment::rbind()</code> returning one valid <code>SummarizedExperiment</code> object.</p></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.48.900" id="jamses-0048900">jamses 0.0.48.900<a class="anchor" aria-label="anchor" href="#jamses-0048900"></a></h2>
<div class="section level3">
<h3 id="silenced-messages-when-loading-jamses-0-0-48-900">silenced messages when loading jamses<a class="anchor" aria-label="anchor" href="#silenced-messages-when-loading-jamses-0-0-48-900"></a></h3>
<ul><li>
<p><code>design</code>, <code>design&lt;-</code> <code>contrasts</code>, <code>contrasts&lt;-</code></p>
<ul><li>Previously when loading <code>jamses</code>, a message indicate that new generic methods were defined, which was caused either by creating methods for an existing function or S4 method that did not have the same signature.</li>
<li>
<code>design</code> now imports and matches the arguments of <code><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">BiocGenerics::design</a></code>
</li>
<li>
<code>contrasts</code> now sets generic functions matching <code><a href="https://rdrr.io/r/stats/contrasts.html" class="external-link">stats::contrasts()</a></code> then adds S4 generics for object <code>SEDesign</code>.</li>
<li>Now, no more warning message when loading <code>jamses</code>!</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.47.900" id="jamses-0047900">jamses 0.0.47.900<a class="anchor" aria-label="anchor" href="#jamses-0047900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-47-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-47-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>
<p>New argument <code>max_correlation_rows</code> with a maximum number of rows to use when calculating <code><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">limma::duplicateCorrelation()</a></code>.</p>
<ul><li>This process is only relevant when <code>block</code> is defined, and <code>correlation=NULL</code>. In this case, <code>limma::lmFit() by default  already calls</code>limma::duplicateCorrelation()<code>to calculate</code>correlation<code>. However, when there are more than 10k rows of data,  the process is non-linearly slow (they said quadratic) although  increasing number of rows does not substantially change the  calculated</code>correlation`.</li>
<li>Therefore <code>max_correlation_rows=10000</code> limits the calculation to 10,000 rows by default, substantially improving speed from 10 minutes in 500k rows, down to 10 seconds.</li>
</ul></li>
<li>
<p>New argument <code>voom_block_twostep=TRUE</code> which enables a <strong>change to previous behavior</strong>.</p>
<ul><li>Only relevant when <code>use_voom=TRUE</code> and <code>block</code> is defined, but <code>correlation</code> is not defined.</li>
<li>Previous behavior was to perform “one-step voom” which calculated weights without the blocking factor, then calculated correlation by <code><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">limma::duplicateCorrelation()</a></code> with Voom weights, and blocking factor.</li>
<li>Recommendation by <code>limma</code> authors is to run these steps twice, so the second time running <code>voom()</code> will use the <code>block</code> and <code>correlation</code> arguments to empower a better model fit, and more accurate <code>weights</code> used subsequently by <code>lmFit()</code>. This change should improve output, but will change previous results, only for count data with a blocking factor, which has been fairly rare.</li>
<li>Previous behavior can be re-enabled with <code>voom_block_twostep=FALSE</code>.</li>
</ul></li>
<li><p>Hid some verbose output.</p></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.46.900" id="jamses-0046900">jamses 0.0.46.900<a class="anchor" aria-label="anchor" href="#jamses-0046900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-46-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-46-900"></a></h3>
<ul><li>
<p><code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>new argument <code>sestats_style</code> to customize how stat hits are labaled. It allows including directionality, only the number of hits, or only the number by itself.</li>
<li>fixed error when any argument of <code>axis1,axis2,axis3,axis4</code> referred to a non-existent column. The error was not a bug, but was avoidable. Now the plot is drawn as if the corresponding argument were <code>NULL</code>.</li>
</ul></li>
<li>
<p><code><a href="../reference/draw_twoway_contrast.html">draw_twoway_contrast()</a></code></p>
<ul><li>new argument <code>contingency</code> to customize how two-way connectors are drawn, based upon how close the two contrasts are to each other. The “loop” option draws the connector to loop around the far side of contrast 2, which is effective at avoiding overlaps.</li>
</ul></li>
<li>
<p><code><a href="../reference/contrast_colors_by_group.html">contrast_colors_by_group()</a></code></p>
<ul><li>
<code>sample_color_list</code> now allows more flexibility in defining colors, specifically allowing a <code>character</code> vector, or single color to be used for all contrasts.</li>
<li>new arguments <code>C_min</code> and <code>L_max</code> to complete the chroma (C) and luminance (L) numerical limits with min/max.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.45.900" id="jamses-0045900">jamses 0.0.45.900<a class="anchor" aria-label="anchor" href="#jamses-0045900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-45-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-45-900"></a></h3>
<ul><li>
<p><code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>So far, early testing is working quite well, minor aesthetic issues, such as unexpected ordering of axis labels when changing the axis order, since it followed the input order and did not re-sort based upon axis label values. Overlapping labels are already an issue. They can be manually adjusted, but it would be nice to have that part automated. Sadly, I don’t think ggrepel accepts rotated text.</li>
<li>change: contrasts are now sorted by axis to organize factor levels</li>
<li>change: axis labels are now sorted by combined factor order along x- and y-axes</li>
<li>change: contrasts are sorted by axis length, so shorter contrasts have less “bump”</li>
<li>fixed bug with missing contrast labels when subsetting the overall contrasts with argument <code>which_contrasts</code>
</li>
<li>fixed handling of <code>colorset</code> to match one-way <code>contrast_names</code> then fill missing color assignments as needed.</li>
<li>arguments <code>group_color</code>,<code>group_border</code> accept named vectors to define colors for each square drawn per experimental group.</li>
</ul></li>
<li>
<p><code><a href="../reference/draw_oneway_contrast.html">draw_oneway_contrast()</a></code>, <code><a href="../reference/draw_twoway_contrast.html">draw_twoway_contrast()</a></code></p>
<ul><li>default <code>label_font=2</code> for bold contrast labels</li>
</ul></li>
<li>
<p><code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code>, <code><a href="../reference/draw_oneway_contrast.html">draw_oneway_contrast()</a></code>, <code><a href="../reference/draw_twoway_contrast.html">draw_twoway_contrast()</a></code></p>
<ul><li>changed argument name from <code>label_position</code> to <code>oneway_position</code> and from <code>twoway_label_position</code> to <code>twoway_position</code>, before these functions are used</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-45-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-45-900"></a></h3>
<ul><li>
<p><code><a href="../reference/contrast_colors_by_group.html">contrast_colors_by_group()</a></code></p>
<ul><li>Takes <code>sedesign</code> and <code>sample_color_list</code> and blends group colors into contrasts based upon the groups in each contrast.</li>
</ul></li>
<li>
<p><code><a href="../reference/sedesign_to_factors.html">sedesign_to_factors()</a></code></p>
<ul><li>Convenience function to convert <code>SEDesign</code> object into the associated <code>data.frame</code> with one column per experimental design factor.</li>
<li>The <code>data.frame</code> rownames match the <code>samples(sedesign)</code>.</li>
<li>The colnames are matched with optional <code>se</code> object in order to re-use <code>colnames(colData(se))</code> and, if present, factor levels. Maintaining factor level order is useful for downstream applications.</li>
<li>This function will be called by <code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code> and should be useful for <code><a href="https://rdrr.io/pkg/platjam/man/design2colors.html" class="external-link">platjam::design2colors()</a></code> as well.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.44.900" id="jamses-0044900">jamses 0.0.44.900<a class="anchor" aria-label="anchor" href="#jamses-0044900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-44-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-44-900"></a></h3>
<p>This release adds <code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code> and several helper functions. This function is still in development, released for more convenient testing for lots of edge cases. This function may be ported to use <code>grid</code> R graphics, but currently uses base R graphics with fixed aspect ratio <code>asp=1</code> to maintain proper visual proportions for block arrows, and group boxes.</p>
<ul><li>
<p><code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>It takes an <code>sedesign</code> object and plots the one-way and two-way contrasts in block arrow format.</li>
<li>Block arrows are “bumped” to reduce directly overlapping arrows.</li>
<li>The design factors can be repositioned along x- and y-axes, at the bottom or top, with combinations of factors, or absence.</li>
<li>It indicates the <code>n=3</code> number of replicates observed per group, and draws shaded squares for groups present, and lack of square for groups which are absent in the design.</li>
<li>When <code>sestats</code> is also supplied, it adds labels with the number of statistical hits per contrast.</li>
<li>Labels can be positioned by one-way or two-way contrast type, or by specific contrast name.</li>
<li>Two-way contrasts can optionally be “flipped” to show the equivalent two-way contrast with different group ordering.</li>
</ul></li>
<li><p><code><a href="../reference/draw_oneway_contrast.html">draw_oneway_contrast()</a></code>, <code><a href="../reference/draw_twoway_contrast.html">draw_twoway_contrast()</a></code> are both called by <code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code>, and internally call <code><a href="../reference/point_handedness.html">point_handedness()</a></code> and <code><a href="../reference/make_block_arrow_polygon.html">make_block_arrow_polygon()</a></code>. These functions help make the process more component-oriented.</p></li>
<li><p><code><a href="../reference/make_block_arrow_polygon.html">make_block_arrow_polygon()</a></code> is a wrapper function to define block arrow polygons with consistent dimensions, where the arrow stem width also defines the relative size of the arrow head. When the stem has shorter length than the arrow head, the full arrow head is shown without the stem, avoiding the visual artifact of showing an inverted (negative length) arrow stem.</p></li>
<li><p><code><a href="../reference/point_handedness.html">point_handedness()</a></code> is a helper function which takes two points and a vector angle, then determines “which side” the first point is located along that vector angle relative to the second point. It is used for two-way contrasts to decide how to define intermediate control points between two parallel contrast arrows. The goal is to create an “S” shape, where control points are angled “toward each other”.</p></li>
<li>
<p><code><a href="../reference/point_slope_intercept.html">point_slope_intercept()</a></code> is a simple geometry helper function.</p>
<ul><li>Given a point and slope, it returns the intercept.</li>
<li>However, when the slope is infinite (as with a vertical line) instead of returning an infinite intercept, which is not helpful for describing the line, it returns the x-axis intercept.</li>
<li>The intercept is used by <code><a href="../reference/point_handedness.html">point_handedness()</a></code> to determine the “handedness” of two parallel lines.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-44-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-44-900"></a></h3>
<ul><li>The method <code>design()</code> now imports the Bioconductor <code>BiocGenerics</code> in order to extend (and thus be indexed alongside) the corresponding method <code>design()</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.43.900" id="jamses-0043900">jamses 0.0.43.900<a class="anchor" aria-label="anchor" href="#jamses-0043900"></a></h2>
<ul><li>Added <code>multienrichjam</code> to dependencies, since it is used by <code><a href="../reference/handle_na_values.html">handle_na_values()</a></code>.</li>
<li>Added <code>statmod</code> to <code>"Suggests"</code> since this package is required when using a blocking factor with <code>limma</code>.</li>
</ul><div class="section level3">
<h3 id="changes-to-existing-functions-0-0-43-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-43-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>The default heatmap title now also includes the number of columns, whose label is controlled by <code>column_type="samples"</code> by default.</li>
<li>When <code>row_type</code> or <code>column_type</code> are <code>NULL</code> or <code>""</code> (empty) their respective label is hidden from the heatmap title. When both are hidden, the first line will be hidden completely.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-43-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-43-900"></a></h3>
<ul><li>
<p><code><a href="../reference/contrast2comp.html">contrast2comp()</a></code></p>
<ul><li>Weird edge case when calling <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code>, <code>sestats</code> contains only one contrast, and the hit matrix annotation contrast is in the form <code>"Group1-Group2"</code>. The output is correct inside RStudio, the hit matrix name displays <code>"Group1-Group2"</code> as intended. When the output is knitted inside an RMarkdown, the hit matrix name is displayed <code>"1"</code>.</li>
<li>Issue was avoided with argument <code>rename_contrasts=FALSE</code> when calling <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code>.</li>
<li>User is using R-3.6.2, so the best guess is a weird edge case where <code>options(stringsAsFactors=TRUE)</code>.</li>
<li>The <code><a href="../reference/contrast2comp.html">contrast2comp()</a></code> was updated to impose <code>stringsAsFactors=FALSE</code> and <code>check.names=FALSE</code> in two steps that converted <code>matrix</code> to <code>data.frame</code>.</li>
<li>The <code><a href="../reference/process_sestats_to_hitim.html">process_sestats_to_hitim()</a></code> was updated to check output of <code><a href="../reference/contrast2comp.html">contrast2comp()</a></code> before assigning to <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code>.</li>
</ul></li>
<li>
<p><code><a href="../reference/handle_na_values.html">handle_na_values()</a></code></p>
<ul><li>Error was thrown due to data returned in <code>list</code> form instead of matrix form. Instead, optional weights are returned as an attribute, so the data is always consistently returned in <code>numeric</code> matrix form. Frankly, it is unclear how this bug was tolerated except calling <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> with <code>handle_na="none"</code>.</li>
<li>The argument <code>na_value</code> was not being passed to <code><a href="../reference/handle_na_values.html">handle_na_values()</a></code> which is important for certain types of data that have high background noise floor values, for example proteomics SpectrumMill has background around 20, also this is in log2 units.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-43-900-1">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-43-900-1"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>now properly passes <code>na_value</code> to <code><a href="../reference/handle_na_values.html">handle_na_values()</a></code>
</li>
</ul></li>
<li>
<p><code><a href="../reference/handle_na_values.html">handle_na_values()</a></code></p>
<ul><li>Nnow returns only <code>numeric</code> matrix, and optionally includes a weight matrix as an attribute named <code>"weights"</code>.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.42.900" id="jamses-0042900">jamses 0.0.42.900<a class="anchor" aria-label="anchor" href="#jamses-0042900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-42-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-42-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_collapse_by_column.html">se_collapse_by_column()</a></code></p>
<ul><li>changed argument default <code>noise_floor_value=NA</code> to <code>noise_floor_value=0</code> so that values are not lost during this process by default. Rare change to default argument, intended to be more consistent with “default” expectations of this function.</li>
<li>new argument <code>useMedian=FALSE</code> ensures the default behavior is to calculate row group mean and not median.</li>
<li>the <code>groupFunc()</code> is now called using <code>jamba::call_fn_ellipse()</code> which ensures that arguments are only passed when permitted by that function, either by named argument, or by <code>...</code>. This change makes it easier to use a custom function which may not have to accept additional arguments.</li>
</ul></li>
<li>
<p><code><a href="../reference/se_collapse_by_row.html">se_collapse_by_row()</a></code></p>
<ul><li>new argument option <code>group_func_name="mean"</code> which enables basic <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code> across group values.</li>
<li>help docs were reworded to clarify the <code>group_func_name</code> and <code>data_transform</code> arguments.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.41.900" id="jamses-0041900">jamses 0.0.41.900<a class="anchor" aria-label="anchor" href="#jamses-0041900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-41-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-41-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>This function adds alternative labels at the top of a rendered <code>Heatmap</code>, sometimes a cleaner alternative than using colored annotation boxes.</li>
<li>The specific scenarios to use this function are described in the help docs, and this function may be updated to be more efficient in future.</li>
<li>Although this function can only be applied after drawing a <code>Heatmap</code>, and requires some blank whitespace inside that heatmap. Nonetheless, this function does make remaining steps substantially more convenient than otherwise possible, and the output looks quite nice for a publication-quality final figure.</li>
</ul></li>
<li>
<p><code><a href="../reference/detect_heatmap_components.html">detect_heatmap_components()</a></code></p>
<ul><li>a helper function intended to be used with <code><a href="../reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code> to identify the <code>grid</code> layout elements defined by the drawn <code>Heatmap</code> in the current display device.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-41-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-41-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>The heatmap title was altered so the centering label uses the pattern <code>"centered within *centerby_colnames*, versus *control_label*"</code>
</li>
<li>help docs were expanded and re-organized to improve clarity.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-41-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-41-900"></a></h3>
<ul><li>
<p><code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code></p>
<ul><li>fixed edge case where columns could be returned as <code>list</code> type, instead of <code>character</code> vector. It resulted in the <code>data.frame</code> appearing to have abbreviated counts, instead of including the full hits, with up- and down-regulated numbers in parentheses.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.40.900" id="jamses-0040900">jamses 0.0.40.900<a class="anchor" aria-label="anchor" href="#jamses-0040900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-40-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-40-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>MAJOR: default argument <code>handle_na="partial"</code> changed to <code>handle_na="full1"</code>. This change is done to “recover” all comparisons where one group has entirely <code>NA</code> values, by substituting one numeric value for one replicate in the group in order to avoid mis-representing the variability of the group. Meanwhile, groups with <code>numeric</code> and <code>NA</code> values together will use only <code>numeric</code> values in the group, without substituting a <code>numeric</code> replacement.</li>
<li>New argument <code>rowData_colnames</code> to retain gene annotation columns for <code>se</code> input where the rownames are not sufficient description of each row. This argument removes the need to merge additional annotation afterward.</li>
<li>Arguments are now fully documented.</li>
<li>handles empty <code>assay_name</code> by returning <code>NULL</code> for each empty assay. Recovers when multiple <code>assay_name</code> values are supplied but some are missing. It does not recover when no assays are found.</li>
<li>Added minor <code>verbose</code> output indicating each <code>assay_name</code>.</li>
</ul></li>
<li>
<p><code><a href="../reference/run_limma_replicate.html">run_limma_replicate()</a></code>, <code><a href="../reference/ebayes2dfs.html">ebayes2dfs()</a></code></p>
<ul><li>New argument <code>rowData_df</code> to support new argument <code>rowData_colnames</code> in <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code>. This argument merges annotations into the resulting top tables so that gene annotations can be retained per row.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.39.900" id="jamses-0039900">jamses 0.0.39.900<a class="anchor" aria-label="anchor" href="#jamses-0039900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-39-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-39-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>visual glitch apparent in version R-4.2.3, the border color for color legend boxes was not black, so new argument <code>legend_border_color</code> was added, and is applied to all legends as relevant.</li>
</ul></li>
<li><p><code><a href="../reference/se_normalize.html">se_normalize()</a></code> - Fixed minor issue with <code>if (class(x))</code> logic to avoid error when comparing multiple values in an <code>if()</code> statement.</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.38.900" id="jamses-0038900">jamses 0.0.38.900<a class="anchor" aria-label="anchor" href="#jamses-0038900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-38-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-38-900"></a></h3>
<ul><li>
<p><code><a href="../reference/contrast_names_to_sedesign.html">contrast_names_to_sedesign()</a></code></p>
<ul><li>convenience function to convert contrast names to corresponding <code>sedesign</code> object.</li>
<li>Its core utility is that it provides a way to test whether two-way or higher contrasts are mathematically equivalent.</li>
<li>It may also be useful for test cases to produce a quick <code>sedesign</code> with only a specific set of contrasts.</li>
<li>In future <code><a href="../reference/plot_sedesign.html">plot_sedesign()</a></code> will display the design with layout similar to <code><a href="https://rdrr.io/pkg/vcd/man/mosaic.html" class="external-link">vcd::mosaic()</a></code>, then will use block arrows to indicate contrasts across groups.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-methods-0-0-38-900">new methods<a class="anchor" aria-label="anchor" href="#new-methods-0-0-38-900"></a></h3>
<ul><li>
<p><code>contrast_names()</code></p>
<ul><li>enables quickly retrieving the vector of contrasts in a <code>sedesign</code> object</li>
<li>also enables re-defining contrasts given a specific vector of contrast names.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="unit-tests-0-0-38-900">unit tests<a class="anchor" aria-label="anchor" href="#unit-tests-0-0-38-900"></a></h3>
<ul><li>
<code><a href="../reference/contrast2comp.html">contrast2comp()</a></code> and <code><a href="../reference/contrast2comp.html">comp2contrast()</a></code> are first.</li>
<li>will provide unit tests for all core functions.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.37.900" id="jamses-0037900">jamses 0.0.37.900<a class="anchor" aria-label="anchor" href="#jamses-0037900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-37-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-37-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>new argument <code>color_floor</code> passed to <code><a href="https://rdrr.io/pkg/colorjam/man/col_div_xf.html" class="external-link">colorjam::col_div_xf()</a></code> which optionally prevents coloring numeric values below the floor, therefore starts the color gradient at this value up to <code>color_max</code>.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.36.900" id="jamses-0036900">jamses 0.0.36.900<a class="anchor" aria-label="anchor" href="#jamses-0036900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-36-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-36-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>
<code>row_split</code> passed as integer value when <code>cluster_rows=FALSE</code> caused an error: “Error: Length or nrow of <code>row_split</code> should be same as nrow of <code>matrix</code>.” The error was caused because rows cannot be split into subgroups in the absence of a row clustering.</li>
<li>The same error for <code>column_split</code> was also present, and was fixed.</li>
<li>Resolution is that <code>row_split</code> or <code>column_split</code> is converted to <code>NULL</code> when there is no respective row or column dendrogram or clustering function.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-36-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-36-900"></a></h3>
<ul><li>
<p><code><a href="../reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>Default format for mgm and mean is “num” instead of “int”, which assumes log2-scaled values that are generally at or less than 20.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>the default gap between row annotations is defined by the ComplexHeatmap option <code>"ROW_ANNO_PADDING"</code>, so the gap between heatmap and annotation is the same as the gap between two annotations. In future the gap width might become a separate option but not just yet.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.35.900" id="jamses-0035900">jamses 0.0.35.900<a class="anchor" aria-label="anchor" href="#jamses-0035900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-35-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-35-900"></a></h3>
<p>Dependency on <code>jamma</code> bumped to version <code>0.0.28.900</code> for new arguments to <code>naControlAction</code>, see below.</p>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li><p>New arguments for data centering: <code>controlFloor</code>, <code>naControlAction</code>, <code>naControlFloor</code></p></li>
<li><p><code>controlFloor</code> is a <code>numeric</code> value for optional noise floor used for control groups during centering. If the control group mean/median value is below <code>controlFloor</code>, it is set to <code>controlFloor</code> to define a minimum baseline for centering. It is mostly useful when the effective noise floor for a platform is above zero.</p></li>
<li><p><code>naControlAction</code> and <code>naControlFloor</code> handle the specific case where all <code>controlSamples</code> have <code>NA</code> values for a particular row. By default, all centered values would become <code>NA</code> thereby making platform measurements disappear from the heatmap. Sometimes it is preferable to define another reference so the non-NA values can be indicated in the heatmap.</p></li>
<li><p>See <code><a href="https://rdrr.io/pkg/jamma/man/centerGeneData.html" class="external-link">jamma::centerGeneData()</a></code> for more information.</p></li>
<li>
<p>The most common alternatives:</p>
<ul><li>
<code>naControlAction="row"</code>: uses the remaining non-NA values from that row</li>
<li>
<code>naControlAction="floor"</code>: assigns the <code>numeric</code> value <code>naControlFloor</code>, which is useful when the reference value should be a known baseline limit of detection.</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="../reference/shrink_df.html">shrink_df()</a></code></p>
<ul><li>Argument <code>numShrinkFunc</code> now includes <code>na.rm=TRUE</code> by default.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.34.900" id="jamses-0034900">jamses 0.0.34.900<a class="anchor" aria-label="anchor" href="#jamses-0034900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-34-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-34-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>new argument <code>use_raster=TRUE</code> which allows turning it off. When R package <code>magick</code> is not installed, on some systems it causes an error when trying to render ComplexHeatmap with <code>use_raster=TRUE</code>. Using <code>use_raster=FALSE</code> avoids the error.</li>
</ul></li>
<li>
<p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code></p>
<ul><li>small update to enforce <code>mixedSortDF(..., honorFactor=TRUE)</code> to ensure that <code>data.frame</code> columns that contain <code>factor</code> values will use the <code>levels</code> to define the proper sort order.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.33.900" id="jamses-0033900">jamses 0.0.33.900<a class="anchor" aria-label="anchor" href="#jamses-0033900"></a></h2>
<div class="section level3">
<h3 id="package-dependencies-0-0-33-900">package dependencies<a class="anchor" aria-label="anchor" href="#package-dependencies-0-0-33-900"></a></h3>
<ul><li>
<code>ComplexHeatmap</code> now requires version <code>2.13.4</code> or higher, and points to the Github repository instead of Bioconductor. This change is required for R-3.6.1 whose matching Bioconductor version 3.10 only offers ComplexHeatmap version 2.2.0, and throws an error with <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> when rendering row annotations.</li>
<li>
<code>jamba</code> dependency now requires version <code>0.0.88.900</code>, to pick up the subtle but important change to <code>mixedSort()</code>
</li>
<li>
<code>colorjam</code> was bumped to version <code>0.0.23.900</code> for consistency.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-33-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-33-900"></a></h3>
<ul><li>
<p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code></p>
<ul><li>
</li></ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.32.900" id="jamses-0032900">jamses 0.0.32.900<a class="anchor" aria-label="anchor" href="#jamses-0032900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-32-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-32-900"></a></h3>
<ul><li>
<code><a href="../reference/se_collapse_by_column.html">se_collapse_by_column()</a></code> which is the equivalent of <code><a href="../reference/se_collapse_by_row.html">se_collapse_by_row()</a></code> except that the default math is much simpler, to perform mean calculation with optional outlier removal prior to calculating the mean.</li>
<li>new experimental function <code><a href="../reference/shrink_df.html">shrink_df()</a></code> as potential drop-in replacement for <code><a href="../reference/shrinkDataFrame.html">shrinkDataFrame()</a></code>. This new function is intended to have simpler processing, with less complicated history than <code><a href="../reference/shrinkDataFrame.html">shrinkDataFrame()</a></code>.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-32-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-32-900"></a></h3>
<ul><li>minor updates to help text for <code><a href="../reference/se_collapse_by_row.html">se_collapse_by_row()</a></code>, added <code>@family</code> to this and other functions.</li>
<li>
<code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> was changed from using <code><a href="https://rdrr.io/pkg/matrixStats/man/rowRanges.html" class="external-link">matrixStats::rowMaxs()</a></code> instead to use <code>apply(x, 1, max, na.rm=TRUE)</code> - due to persistent Segfaults caused by <code><a href="https://rdrr.io/pkg/matrixStats/man/rowRanges.html" class="external-link">matrixStats::rowMaxs()</a></code> on multiple architectures (OSX, linux) and with recent package versions, at least for R-3.6.1.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.31.900" id="jamses-0031900">jamses 0.0.31.900<a class="anchor" aria-label="anchor" href="#jamses-0031900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-31-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-31-900"></a></h3>
<ul><li>
<code><a href="../reference/contrast2comp.html">names_contrast2comp()</a></code> and <code><a href="../reference/contrast2comp.html">names_comp2contrast()</a></code> are analogous to <code><a href="../reference/contrast2comp.html">contrast2comp()</a></code> and <code><a href="../reference/contrast2comp.html">comp2contrast()</a></code> except they operate on the object names, making it convenient to wrapper around <code>list</code> objects where the list names are the contrasts or comps.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-31-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-31-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>argument <code>normgroup</code> now accepts <code>colnames(colData(se))</code> to define normgroups by column annotations.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-31-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-31-900"></a></h3>
<ul><li>
<p>fixed apparent regression in <code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> for interaction contrasts</p>
<ul><li>version 0.0.30.900 apparently introduced regression bug in interaction contrasts that returns only the interactions in <code>list</code> form instead of all contrasts in <code>sedesign</code> form. Short-lived, fortunately.</li>
<li>Some debug potential was added with <code>verbose=TRUE</code>, that prints iterative contrast name generation, with indent to help show when the function is iteratively calling itself.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>fixed bug with user-supplied <code>row_names_gp</code> or <code>column_names_gp</code> that does not also define <code>fontsize</code>.</li>
<li>new arguments <code>cluster_column_slices=FALSE</code>, <code>cluster_row_slices=FALSE</code> to allow this behavior to be overridden when relevant</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-31-900-1">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-31-900-1"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>moved substantial logic out, and into <code><a href="../reference/process_sestats_to_hitim.html">process_sestats_to_hitim()</a></code>, which handled the hit incidence matrix, and deciding which rows are included in the heatmap with default <code>rows=NULL</code>.</li>
<li>quality of life change: when there is only one <code>assayNames(se)</code>, then <code>assay_name</code> defaults to the only possible value, instead of acting like it couldn’t possibly know what I mean. Haha.</li>
<li>
<code>legend_at</code> is smarter when <code>centerby_colnames=FALSE</code> and there is no data centering, the legend only includes negative values when data contains negative values.</li>
<li>
<code>legend_labels</code> are smarter when <code>color_max &lt;= 1</code>, it uses the numerical value without transformation.</li>
<li>vastly expanded the examples and the param help text.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-31-900-1">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-31-900-1"></a></h3>
<ul><li>
<p><code><a href="../reference/process_sestats_to_hitim.html">process_sestats_to_hitim()</a></code></p>
<ul><li>previously internal logic to <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code>, it surpassed the amount of logic hidden inside a function, especially when it’s used twice.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.30.900" id="jamses-0030900">jamses 0.0.30.900<a class="anchor" aria-label="anchor" href="#jamses-0030900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-30-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-30-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>removed duplicated <code><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge()</a></code> called across stats <code>data.frame</code> results.</li>
<li>implemented <code>normgroups</code> for independent model fit among subsets of samples.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.29.900" id="jamses-0029900">jamses 0.0.29.900<a class="anchor" aria-label="anchor" href="#jamses-0029900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-29-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-29-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> did not properly check <code>cluster_columns</code> for <code>function</code> input, so it did not allow using a custom clustering function.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-29-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-29-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>new arguments: <code>row_names_gp</code> and <code>column_names_gp</code> allow custom settings for row and column labels, respectively.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.28.900" id="jamses-0028900">jamses 0.0.28.900<a class="anchor" aria-label="anchor" href="#jamses-0028900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-28-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-28-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> with <code>posthoc_test="DEqMS"</code> and only one contrast threw an error <code>"Error in fit$sca.t[results.table$gene, coef_col] : subscript out of bounds"</code>. The error actually occurs in <code><a href="../reference/ebayes2dfs.html">ebayes2dfs()</a></code> the function that converts statistical model fit into a list of <code>data.frame</code> objects.</p>
<ul><li>The <code>DEqMS</code> package produces <code>numeric</code> matrix results that it also summarizes into a <code>data.frame</code> via <code>DEqMS::outputResult()</code>. When there is only one contrast, many of the <code>numeric</code> matrix objects also have one column, and R apparently dropped the <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code> for the <code>sca.t</code> matrix, causing <code>DEqMS::outputResult()</code> to fail when argument <code>coef_col</code> is given the <code>character</code> name of the contrast instead of the <code>integer</code> column number.</li>
<li>The contrast name is preserved in the <code>coefficients</code> matrix, so the contrast is converted to column number, then passed as <code>coef_col</code>, and resolves the error.</li>
<li>Calls to <code>DEqMS::outputResult()</code> use <code>integer</code> values for argument <code>coef_col</code>.</li>
<li>Some internal step in <code>DEqMS</code> does not properly handle data with only one contrast, probably forgot to include <code>drop=FALSE</code> in matrix subsetting.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-28-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-28-900"></a></h3>
<ul><li>
<p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> updates</p>
<ul><li>input <code>SummarizedExperiment</code> is now recognized, using <code>group_colnames</code> to define the experimental design <code>data.frame</code>. Data can also be subset using <code>isamples</code> for example.</li>
<li>new argument <code>contrast_names</code> to provide specific contrasts upfront, therefore skipping all internal logic of defining contrasts.</li>
<li>internal values <code>max_depth</code> and <code>factor_order</code> are validated using the design <code>data.frame</code> such that it will not pursue <code>max_depth</code> nor <code>factor_order</code> higher than the number of design columns available.</li>
<li>When there are no valid contrasts for the given design groups, the function returns empty contrasts instead of throwing an error. One could argue that an error is actually preferred, since it forces the input data to be valid, and forces the user to review the data.</li>
</ul></li>
<li><p><code><a href="../reference/contrast2comp.html">contrast2comp()</a></code> and <code><a href="../reference/contrast2comp.html">comp2contrast()</a></code> were reworked to handle alternate delimiters, which mainly involved wrapping the delimiter in proper regular expression patterns. During the process, obscure testing showed certain delimiters cause problems with regular expressions, so those problems were pre-emptively fixed. The help docs were expanded somewhat.</p></li>
<li><p><code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code> was refactored to be cleaner, correcting the issue of colname mislabling when <code>dimname_order</code> was different than expected. The function now works smoothly with any <code>dimname_order</code> values, and accepts <code>character</code> or <code>integer</code> input to specify the dimname ordering.</p></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-28-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-28-900"></a></h3>
<ul><li>
<p><code><a href="../reference/format_hits.html">format_hits()</a></code> is an internal function, but it might be a useful convenience function so it is exported. It is called by <code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code> to provide a summary <code>data.frame</code> of statistical results. It converts a hit vector into either:</p>
<ul><li>a summary string with hits/up/down,</li>
<li>one integer count of hits, or</li>
<li>a vector of integers for c(“hits”, “up”, “down”)`.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.27.900" id="jamses-0027900">jamses 0.0.27.900<a class="anchor" aria-label="anchor" href="#jamses-0027900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-27-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-27-900"></a></h3>
<ul><li>
<p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> threw an error in some circumstances when the input <code>ifactors</code> only contained one column, but the column contained levels with delimiters recognized by <code>factor_sep</code>, and some levels without the delimiter. The error was: <code>"Error in combn(iMatch, 2) : n &lt; m"</code></p>
<ul><li>The workaround was to supply levels with consistent delimiters, or a design that used more than one column initially.</li>
<li>However, the R code was adjusted to handle these situations properly without workarounds.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.26.900" id="jamses-0026900">jamses 0.0.26.900<a class="anchor" aria-label="anchor" href="#jamses-0026900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-26-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-26-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> updates:</p>
<ul><li>still in progress: several new arguments intended to customize exact font size and grid size for color legends, annotations, around the heatmap.</li>
<li>new arguments: <code>show_heatmap_legend</code>, <code>show_top_legend</code>, <code>show_left_legend</code> intended to allow hiding various color legends, mostly to save plot space.</li>
<li>
<code>all_sample_colors</code> sometimes does not match left or top annotation names, and now will create categorical colors in more scenarios.</li>
<li>new argument <code>mark_rows</code> to enable optional <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_mark.html" class="external-link">ComplexHeatmap::anno_mark()</a></code> row labels for a subset of heatmap rows. <code>mark_labels_gp</code> to customize the displayed font.</li>
<li>new argument <code>right_annotation</code> for custom annotation</li>
<li>new arguments <code>show_top_annotation_name</code>, <code>show_left_annotation_name</code>, <code>left_annotation_name_rot</code> to customize these label positions.</li>
<li>new arguments <code>mark_rows</code> and <code>mark_labels_gp</code> used to define <code>anno_mark()</code> call-out labels for a subset of rows in the heatmap. This step is difficult to define upfront when the data in <code>se</code> is subset by some other aspect of <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code>, since <code>anno_mark()</code> requires numeric index positions for each label, and the order of rows may not be known upfront.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="minor-bug-fixes-0-0-26-900">minor bug fixes<a class="anchor" aria-label="anchor" href="#minor-bug-fixes-0-0-26-900"></a></h3>
<ul><li>
<code><a href="../reference/hit_array_to_list.html">hit_array_to_list()</a></code> was returning <code>NA</code> values when the <code>sestats$hit_array</code> also included <code>NA</code> values. The function now removes any <code>NA</code> values before returning each vector.</li>
<li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> argument <code>alt_sestats</code> was throwing an error when supplied with a <code>numeric</code> matrix instead of <code>sestats</code> object, this input type was been corrected, and the code slightly refactored for clarity.</li>
<li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was passing multiple values to the left_annotation argument <code>show_legend</code>, which only accepts one <code>logical</code> value. This bug was corrected.</li>
</ul></div>
<div class="section level3">
<h3 id="new-dev-function-0-0-26-900">new dev function<a class="anchor" aria-label="anchor" href="#new-dev-function-0-0-26-900"></a></h3>
<ul><li>
<code>contrasts2comp_dev()</code> is an experiment in converting the code to be vectorized, though the exceptions to handle unbalanced contrasts may still prevent this function from being as fast as intended.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.25.900" id="jamses-0025900">jamses 0.0.25.900<a class="anchor" aria-label="anchor" href="#jamses-0025900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-25-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-25-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> updates:</p>
<ul><li>new argument <code>data_type</code> used in color legend title, and heatmap title.</li>
<li>new arguments <code>legend_at</code>, <code>legend_labels</code> to customize heatmap color gradient legend positions and labels.</li>
<li>data centering can be turned off with <code>centerby_colnames=FALSE</code>.</li>
<li>
<code>sestats</code> and <code>alt_sestats</code> can be supplied as incidence matrices directly, without requiring them to be in <code>sestats</code> or <code>hit_array</code> formats.</li>
<li>
<code>top_colnames</code> can be hidden with <code>top_colnames=FALSE</code>; also when <code>colData(se)</code> is empty, <code>top_annotation</code> is hidden.</li>
<li>
<code>assay_name</code> may contain multiple values to define gene hits in <code>sestats</code>; however only the first matching <code>assay_name</code> in <code>names(assays(se))</code> is used for the heatmap data.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.24.900" id="jamses-0024900">jamses 0.0.24.900<a class="anchor" aria-label="anchor" href="#jamses-0024900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-24-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-24-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> new argument <code>row_subcluster</code>:</p>
<ul><li><p>intended to help drill-down into a subset of heatmap rows by using elements of <code>row_split</code> as returned by <code><a href="https://rdrr.io/pkg/jamba/man/heatmap_row_order.html" class="external-link">jamba::heatmap_row_order()</a></code>.</p></li>
<li>
<p>The rows are split by <code>row_split</code> using the full heatmap data, and can be split by…</p>
<ul><li>cluster number using <code>row_split=5</code> integer values,</li>
<li>annotation columns in <code>rowData(se)</code>,</li>
<li>
<code>character</code> vector, or</li>
<li><code>data.frame</code></li>
</ul></li>
<li><p>In any case, <code>heatmap_row_order()</code> returns a <code>list</code> of <code>rownames(se)</code>, which is used to create the heatmap. In fact, rows are displayed in the same order, but without any associated dendrogram even if there is a dendrogram associated with the subset of rows.</p></li>
<li><p><code>row_subcluster</code> can be an <code>integer</code> vector referring to the row list element in order, or <code>character</code> vector of one or more <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code> associated with the row list elements.</p></li>
</ul></li>
<li><p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> arguments <code>cutoff_name</code> and <code>alt_cutoff_name</code> default values changed to <code>NULL</code>, so hits in <code>sestats</code> and <code>alt_sestats</code> will use all cutoffs by default, unless specified otherwise.</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.23.900" id="jamses-0023900">jamses 0.0.23.900<a class="anchor" aria-label="anchor" href="#jamses-0023900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-23-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-23-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was updated</p>
<ul><li><p><code>isamples</code> is applied to subset the columns displayed in the heatmap, however it no longer subsets data during the data centering step. This change allows centering by patient with time 0 as the control point, which produces a stripe of <code>0</code> values for each patient at time zero, all other times become difference from time zero. The heatmap can then use <code>isamples</code> with only samples not at time zero.</p></li>
<li>
<p>new argument <code>subset_legend_colors=TRUE</code> will now filter colors in the <code>top_annotation</code> and <code>left_annotation</code> color legends to remove categorical colors which are not represented in the data being displayed.</p>
<ul><li>This update prevents showing a huge list of categorical colors that are not relevant to the heatmap. Note that even when two heatmaps are added together (with ComplexHeatmap <code>+</code> or <code>%v%</code>) the amazing coding in ComplexHeatmap will merge together color legends by annotation color name, showing only the unique set of colors.</li>
<li>It seems that ComplexHeatmap retains all colors in the heatmap object, but hides colors that are unused. The <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> code was actually encoding the <code>at</code> color labels for all colors regardless of which colors were present. So it is unclear the benefit of this option, but the function itself was improved.</li>
</ul></li>
<li><p>color legends are displayed either in order of factor levels, or sorted using <code><a href="https://rdrr.io/pkg/jamba/man/mixedSort.html" class="external-link">jamba::mixedSort()</a></code></p></li>
<li><p>new argument <code>row_anno_fontsize</code> to control the specific fontsize for row annotation labels. These labels appear for row annotations, and appears beside the column labels on the heatmap.</p></li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-23-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-23-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> when <code>sestats</code> is supplied but <code>se</code> does not contain all rows present in <code>sestats</code> hit array, it produced an error.</li>
<li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> when annotations with factor columns were subset to remove a color from the legend, the color legend was still encoding <code>at</code> labels for the full set of colors, causing <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">ComplexHeatmap::HeatmapAnnotation()</a></code> to print a warning message.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-23-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-23-900"></a></h3>
<ul><li>
<code><a href="../reference/choose_annotation_colnames.html">choose_annotation_colnames()</a></code> is a wrapper function to hold the simple logic for choosing annotation columns to include for <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code>. By default it chooses columns with at least one repeated value, at least two unique values, and only includes the first column with matching cardinality. The last criterion ensures it won’t display six columns which show the same pattern of information for all samples.</li>
</ul></div>
<div class="section level3">
<h3 id="removed-functions-0-0-23-900">removed functions<a class="anchor" aria-label="anchor" href="#removed-functions-0-0-23-900"></a></h3>
<ul><li>
<code>call_fn_ellipsis()</code> was removed since it was migrated to <code>jamba</code>. (Previous change.)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.22.900" id="jamses-0022900">jamses 0.0.22.900<a class="anchor" aria-label="anchor" href="#jamses-0022900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-22-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-22-900"></a></h3>
<ul><li>
<code><a href="../reference/hit_array_to_list.html">hit_array_to_list()</a></code> was returning <code>matrix</code> for single contrasts, due to <code><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply()</a></code> tendency to simplify output to vector. The method was changed to use a different approach guaranteed to return <code>list</code>.</li>
<li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> had a typo in an edge condition when using <code>alt_sestats</code>, which has been corrected.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.21.900" id="jamses-0021900">jamses 0.0.21.900<a class="anchor" aria-label="anchor" href="#jamses-0021900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-21-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-21-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was not properly auto-detecting <code>show_row_names</code>, which only worked when <code>rows</code> was supplied specifically. Now it works when <code>rows</code> is not supplied and/or only <code>sestats</code> is also supplied.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.20.900" id="jamses-0020900">jamses 0.0.20.900<a class="anchor" aria-label="anchor" href="#jamses-0020900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-20-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-20-900"></a></h3>
<ul><li>
<p><code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code> was updated:</p>
<ul><li>now reports <code>""</code> whenever the <code>hit_array</code> value is NA, because the NA in this context means the cutoff was not tested for statistical hits. This situation happens when using <code>int_adjp_cutoff</code>,<code>int_fold_cutoff</code>,<code>int_mgm_cutoff</code> that differ from the main contrast cutoffs. Therefore, reporting <code>0</code> implies the threshold was tested and no hits were found, which is incorrect.</li>
<li>also whenever there are zero hits it reports <code>"0 hits"</code> and not <code>"0 hits (0 up, 0 down)"</code>. This change saves unnecessary text, also makes it easier to see which entries have hits.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.19.900" id="jamses-0019900">jamses 0.0.19.900<a class="anchor" aria-label="anchor" href="#jamses-0019900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-19-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-19-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was updated:</p>
<ul><li>The logic of obtaining a hit list for a contrast must account for multiple cutoffs, particularly when interaction contrast cutoffs differ from pairwise cutoffs. This logic was pushed into new function <code><a href="../reference/hit_array_to_list.html">hit_array_to_list()</a></code>.</li>
<li>The default color legend for <code>left_annotation</code> and <code>top_annotation</code> now includes <code>color_bar="discrete"</code> for numeric annotation columns. The legend will display discrete color breaks instead of a continuous color bar. The breaks are defined in the color function, in most cases the recommended method is calling <code><a href="https://rdrr.io/pkg/platjam/man/design2colors.html" class="external-link">platjam::design2colors()</a></code>, which typically defines reasonable color breaks.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-19-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-19-900"></a></h3>
<ul><li>
<p><code><a href="../reference/hit_array_to_list.html">hit_array_to_list()</a></code></p>
<ul><li><p>Input is <code>hit_array</code> from <code>sestats</code>, as output from <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code>.</p></li>
<li>
<p>It optionally takes a subset of the following, which match the <code><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames()</a></code> in <code>hit_array</code>:</p>
<ul><li><code>cutoff_names</code></li>
<li><code>contrast_names</code></li>
<li><code>assay_names</code></li>
</ul></li>
<li>
<p>Then it iterates each <code>contrast_names</code>:</p>
<ul><li>It returns one named vector of hits for that contrast, named by the measurement, with values <code>-1</code> or <code>1</code> indicating the direction.</li>
<li>When multiple <code>assay_names</code> or <code>cutoff_names</code> are involved, it combines results into one set representing the union of hits.</li>
</ul></li>
<li><p>The intent is to retrieve a <code>list</code> of hits across <code>contrast_names</code>, regardless if there is one or more <code>cutoff_names</code> or <code>assay_names</code>.</p></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.18.900" id="jamses-0018900">jamses 0.0.18.900<a class="anchor" aria-label="anchor" href="#jamses-0018900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-18-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-18-900"></a></h3>
<ul><li>
<p><code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code> was updated:</p>
<ul><li>
<p>when <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> uses different interaction statistical thresholds from the pairwise thresholds, the output <code>hit_array</code> contains NA values.</p>
<ul><li>previously any non-empty vector simply counted the length</li>
<li>new behavior is to count non-NA entries</li>
<li>future “fix” will be for the cell to have length=0, with no <code>NA</code>. This situation only happens when <code>*_cutoff</code> and <code>int_*_cutoff</code> define different thresholds.</li>
</ul></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.17.900" id="jamses-0017900">jamses 0.0.17.900<a class="anchor" aria-label="anchor" href="#jamses-0017900"></a></h2>
<div class="section level3">
<h3 id="functions-removed-0-0-17-900">functions removed<a class="anchor" aria-label="anchor" href="#functions-removed-0-0-17-900"></a></h3>
<ul><li>
<code>call_fn_ellipse()</code> was removed and placed in the <code>jamba</code> package, package dependency was bumped up to <code>jamba(&gt;= 0.0.81.900)</code>
</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-17-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-17-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> error when not supplied <code>rows</code> nor <code>sestats</code>, now defaults to show all <code>rownames(se)</code>.</li>
<li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> error when <code>colorjam</code> package not previously loaded, added missing package prefix.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-17-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-17-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> argument <code>normgroup_colname=NULL</code> and <code>centerby_colnames=NULL</code>, instead of previous defaults which should not be used unless relevant to the design.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-17-900-1">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-17-900-1"></a></h3>
<ul><li>
<code><a href="../reference/save_sestats.html">save_sestats()</a></code> was updated to prioritize sorting <code>row_type</code> column headers first in the output Excel table, previously the method looked for known patterns, which obviously failed when <code>row_type</code> differed from those known patterns. Now the <code>row_type</code> value itself is used in addition to the expected patterns.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.16.900" id="jamses-0016900">jamses 0.0.16.900<a class="anchor" aria-label="anchor" href="#jamses-0016900"></a></h2>
<div class="section level3">
<h3 id="functions-removed-0-0-16-900">functions removed<a class="anchor" aria-label="anchor" href="#functions-removed-0-0-16-900"></a></h3>
<ul><li><code>call_fn_ellipse()</code></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-16-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-16-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> new argument <code>correlation</code></p>
<ul><li>when <code>correlation=TRUE</code> it calculates a sample correlation matrix, and will display a correlation heatmap.</li>
<li>The option was added to this function in order to share much of the remaining logic involving data centering, and heatmap column annotations.</li>
<li>Note that it is possible to create a correlation heatmap using a subset of statistical hits, which although typically creates a nice-looking correlation structure, may not be that informative since it only includes statistical hits.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-16-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-16-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> error when <code>row_split</code> was an integer number of clusters, and <code>cluster_rows</code> is a <code>function</code>, which is not permitted by <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>. Instead, when <code>cluster_rows</code> or <code>cluster_columns</code> is a function, the function is evaluated upfront, so the resulting dendrogram or hclust is passed to <code>Heatmap()</code>, which does permit integer row and column split.</p>
<ul><li>Note that when <code>row_split</code> or <code>column_split</code> are not single numeric values, the <code>function</code> must not be evaluated, otherwise that process fails. Haha. Makes sense. Now both scenarios are handled, hopefully this covers all scenarios.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.15.900" id="jamses-0015900">jamses 0.0.15.900<a class="anchor" aria-label="anchor" href="#jamses-0015900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-15-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-15-900"></a></h3>
<ul><li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Argument <code>row_split</code> now also accepts named vector where names correspond to rownames in the heatmap.</li>
<li>Added help text for all function arguments.</li>
</ul></li>
<li>
<p>save_sestats()`</p>
<ul><li>new argument <code>use_assay_suffix=TRUE</code> will try to add abbreviated <code>assay_names</code> to the end of each Excel sheet name, when there is more than one unique <code>assay_names</code> value.</li>
<li>The Excel sheet names are required to be unique, and have no more than <code>max_nchar_sheetname</code> characters in each string. The sheet names are progressively shortened as necessary, then are made unique when required by calling <code><a href="https://rdrr.io/pkg/jamba/man/makeNames.html" class="external-link">jamba::makeNames()</a></code>.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.14.900" id="jamses-0014900">jamses 0.0.14.900<a class="anchor" aria-label="anchor" href="#jamses-0014900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-14-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-14-900"></a></h3>
<p>This update attempts to extend <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> with methods from the Bioconductor package DEqMS. This package models error as a function of PSM_count, which is the number of peptide score matrix entries that were combined to formulate an abundance estimate for each protein. The approach is analogous to using <code>voom</code> as an extension of <code>limma</code>. In fact, DEqMS is another extension of the <code>limma</code> error model.</p>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>argument <code>posthoc_test=c("none", "DEqMS")</code> to allow post-hoc tests such as <code>DEqMS</code> to be applied after the core <code>limma</code> steps. This argument is pushed into <code>run_limma_replicates()</code>.</li>
<li>argument <code>posthoc_args</code> is a list named by the <code>posthoc_test</code> value, containing a list of named elements used as arguments for the relevant post-hoc test functions.</li>
</ul></li>
<li>
<p><code><a href="../reference/run_limma_replicate.html">run_limma_replicate()</a></code></p>
<ul><li>argument <code>posthoc_test=c("none", "DEqMS")</code> to allow post-hoc tests such as <code>DEqMS</code> to be applied after the core <code>limma</code> steps. This argument is also passed to <code><a href="../reference/ebayes2dfs.html">ebayes2dfs()</a></code>.</li>
<li>argument <code>posthoc_args</code> is a list named by the <code>posthoc_test</code> value, containing a list of named elements used as arguments for the relevant post-hoc test functions.</li>
<li>argument <code>trim_colnames</code> added <code>"sca.t"</code> as additional column to omit by default.</li>
</ul></li>
<li>
<p><code><a href="../reference/ebayes2dfs.html">ebayes2dfs()</a></code></p>
<ul><li>argument <code>lmFit4</code> is optional and intended to contain additional output from post-hoc methods such as <code>DEqMS::spectraCounteBayes()</code>.</li>
<li>argument <code>trim_colnames</code> added <code>"sca.t"</code> as additional column to omit by default.</li>
<li>The order of colnames in the <code>data.frame</code> returned was updated to be more consistent for different formats. Specifically, when <code>posthoc_test="DEqMS"</code> the P-value columns sort with the <code>"sca.P.Value"</code> and <code>"sca.adj.pval"</code> columns before the usual limma columns.</li>
<li>when calling <code><a href="../reference/mark_stat_hits.html">mark_stat_hits()</a></code> it updates the expected colnames when <code>posthoc_test="DEqMS"</code>, using <code>adjp_colname="sca.adj.pval"</code> and <code>p_colname="sca.P.Value"</code>. In this way, hits are filtered by the DEqMS adjusted P-value and P-value, respectively. All other statistics are carried over from <code>lmFit3</code> and work as before.</li>
</ul></li>
<li>
<p><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>The incidence matrix data for <code>alt_sestats</code> was updated to handle proper hit matrix, in haste to make the <code>left_annotation</code> extensible the sestats hit matrix was used twice. This is why the package is private, for now.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.13.900" id="jamses-0013900">jamses 0.0.13.900<a class="anchor" aria-label="anchor" href="#jamses-0013900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-13-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-13-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was modified to handle updated <code>sample_color_list</code> that contains <code>function</code> output from <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">circlize::colorRamp2()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.12.900" id="jamses-0012900">jamses 0.0.12.900<a class="anchor" aria-label="anchor" href="#jamses-0012900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-12-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-12-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> was modified to accomodate other object types, such as <code>"MethyLumiSet"</code> which does not have <code>rowData()</code>,<code>colData()</code>,<code>assays()</code>, accessor functions, instead uses a common alternative for Bioconductor objects: <code>featureData()</code>, <code>phenoData()</code> and <code>assayData()</code>. The <code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> changes will try to use those accessor functions for any object whose class does not contain <code>"SummarizedExperiment"</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.11.900" id="jamses-0011900">jamses 0.0.11.900<a class="anchor" aria-label="anchor" href="#jamses-0011900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-11-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-11-900"></a></h3>
<ul><li>
<code><a href="../reference/se_detected_rows.html">se_detected_rows()</a></code> is a new function to assist the process of defining “detected” rows which are suitable for downstream analysis. The rules are essentially simple heuristics based upon the total number of replicates, group replicates, fraction of group replicates, and number of sample groups where a row has “valid” measurements. In this case “valid” is defined as meeting a minimum abundance threshold, and is not <code>NA</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.10.900" id="jamses-0010900">jamses 0.0.10.900<a class="anchor" aria-label="anchor" href="#jamses-0010900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-10-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-10-900"></a></h3>
<ul><li>
<p><code><a href="../reference/contrast2comp.html">contrast2comp()</a></code>, <code><a href="../reference/contrast2comp.html">comp2contrast()</a></code> are two reciprocal functions intended to convert long contrast names to a short comparison form called a <code>"comp"</code>. The <code>"comp"</code> can be converted back to the full contrast name. In cases where a two-way contrast can be written in two equivalent forms, argument <code>factor_order</code> can be used to define the desired order. The statistical results are the same, but sometimes it helps to have specific ordering. See help docs and examples for details, and worked examples. Main driving motivation:</p>
<ul><li><p><code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code> labels are too long to fit onscreen without making other non-ideal adjustments to plot dimensions.</p></li>
<li><p><code><a href="https://rdrr.io/pkg/venndir/man/venndir.html" class="external-link">venndir::venndir()</a></code> labels are also too long.</p></li>
<li>
<p>Excel worksheets are limited to 31 characters, a typical two-way contrast with 3-characters for each factor level uses 35 characters, the <code>"comp"</code> for uses only 15.</p>
<ul><li>contrast: <code>"(Aaa_Bbb-Ccc_Bbb)-(Aaa_Ddd-Ccc_Ddd)"</code> (35 characters)</li>
<li>comp: <code>"Aaa-Cca:Bba-Dda"</code> (15 characters)</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-10-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-10-900"></a></h3>
<ul><li>
<p>argument <code>rename_contrasts</code> was added to:</p>
<ul><li><code><a href="../reference/heatmap_se.html">heatmap_se()</a></code></li>
<li><code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code></li>
<li><code><a href="../reference/save_sestats.html">save_sestats()</a></code></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.9.900" id="jamses-009900">jamses 0.0.9.900<a class="anchor" aria-label="anchor" href="#jamses-009900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-9-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-9-900"></a></h3>
<ul><li>
<code><a href="../reference/save_sestats.html">save_sestats()</a></code> was updated to use <code>jamba-version-0.0.79-900</code> new <code>writeOpenxlsx()</code>, which uses a substantially faster method of saving worksheets, and allows building a multi-sheet Workbook prior to saving to a file. This change avoids the time-consuming, repeated steps: loading, saving, updating, saving, loading, updating, saving, etc. Previously, each successive worksheet added to an Excel file took longer to load and save. For a 15,000 row, 8 column <code>data.frame</code>, saving once took 3 seconds, but saving 12 worksheets took 15 minutes! The new approach took about 25 seconds.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-9-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-9-900"></a></h3>
<ul><li>
<code><a href="../reference/heatmap_se.html">heatmap_se()</a></code> is a work in progress, but attempts to automate display of expression data as a heatmap using <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>. By default it uses statistical hits defined in <code>sestats</code> output from <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code>.</li>
<li>
<code>call_fn_ellipsis()</code> is copied from <code>multienrichjam::call_fn_ellipsis()</code>. This function allows passing ellipsis argument <code>...</code> to another function, where that function does not allow <code>...</code> and therefore all arguments must be strictly defined. Any arguments in <code>...</code> which are not defined in the other function are removed before calling that function.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.8.900" id="jamses-008900">jamses 0.0.8.900<a class="anchor" aria-label="anchor" href="#jamses-008900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-8-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-8-900"></a></h3>
<ul><li>
<p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> new behavior</p>
<ul><li>two-way interaction contrasts now retain the order of first-order contrasts, preferring to alter the reverse <code>rev(factor_order)</code> in order to build interaction contrasts on top of existing contrasts.</li>
<li>some verbose output now requires <code>(verbose &gt;= 2)</code>
</li>
<li>added verbose output for two-way contrasts removed because they are equivalent, but written with different order of comparisons.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.7.900" id="jamses-007900">jamses 0.0.7.900<a class="anchor" aria-label="anchor" href="#jamses-007900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-7-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-7-900"></a></h3>
<ul><li>
<p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> new argument <code>default_sort=c("asis", "sort_samples", "mixedSort")</code>:</p>
<ul><li><p>Note in all cases, a <code>factor</code> column is sorted by its factor levels.</p></li>
<li><p><code>"asis"</code>: will convert any non-<code>factor</code> column to <code>factor</code>, with levels defined in the order they appear in the input data. Any existing <code>factor</code> column will remain as-is.</p></li>
<li><p><code>"sort_samples"</code>: will call <code><a href="../reference/sort_samples.html">sort_samples()</a></code> in order to recognize control terms, and preferentially place them first, for example “Control”, or “Wildtype”. All other terms are sorted by <code><a href="https://rdrr.io/pkg/jamba/man/mixedSort.html" class="external-link">jamba::mixedSort()</a></code> for alphanumeric sorting.</p></li>
<li>
<p><code>"mixedSort"</code>: will use <code><a href="https://rdrr.io/pkg/jamba/man/mixedSortDF.html" class="external-link">jamba::mixedSortDF()</a></code> to sort group columns:</p>
<ul><li>
<code>factor</code> columns are sorted according to factor levels</li>
<li>
<code>character</code> columns are alphanumeric sorted using <code><a href="https://rdrr.io/pkg/jamba/man/mixedSort.html" class="external-link">jamba::mixedSort()</a></code>
</li>
</ul></li>
<li><p>To influence the sort order, <code>data.frame</code> columns should have <code>factor</code> values with defined levels in order. Otherwise by default, the first entries in the <code>data.frame</code> provided will become the first factor levels.</p></li>
</ul></li>
<li><p><code><a href="../reference/validate_sedesign.html">validate_sedesign()</a></code> default argument <code>verbose=FALSE</code>, also silencing other internal messages which were previously always turned on.</p></li>
<li>
<p><code><a href="../reference/matrix_normalize.html">matrix_normalize()</a></code> new argument <code>normgroup</code> normalizes subsets of matrix columns independently, then re-assembles the original matrix.</p>
<ul><li><p>Purpose is convenience of maintaining one data matrix, while also applying normalization independently on each sub-matrix. The attributes returned from the normalization method are also combined, for example the normalization factors <code>"nf"</code> reported by <code><a href="https://rdrr.io/pkg/jamma/man/jammanorm.html" class="external-link">jamma::jammanorm()</a></code> are combined in proper order.</p></li>
<li>
<p>This feature is distinct from batch adjustment, in that the data in each sub-matrix are never normalized relative to each other.</p>
<ul><li>For example, one could normalize total RNA-seq and nascent 4sU-seq data independently, without expectation that the two would ever have a common frame of reference to normalize one relative to another.</li>
<li>Similarly, one could normalize each tissue type independently, which may be appropriate when analyzing data that contains very different mammalian tissue organ samples, such as muscle and brain. It would generally not be appropriate to use quantile normalization across muscle and brain samples, since the overall pattern and distribution of expression values is not expected to be similar. Quantile normalize assumes (and imposes) a common distribution, by adjusting mean expression signal at each quantile to a common mean expression across all samples.</li>
<li>For a rough approximation of cross-tissue normalization, one could apply <code>"quantile"</code> normalization within each <code>normgroup</code> defined by tissue type, then apply <code>"jammanorm"</code> median normalization to apply a linear adjustment of signal across tissue types. The median normalization does not affect distribution, thus will not affect intra-tissue contrasts, except by adjusting its overall signal which may change downstream assumptions regarding signal thresholds. It is still not advise to compare directly across tissue types, except in some cases a two-way contrast may be appropriate (comparing intra-tissue fold change to intra-tissue fold change.)</li>
</ul></li>
</ul></li>
<li><p><code><a href="../reference/se_normalize.html">se_normalize()</a></code> also gained new argument <code>normgroup</code>, although this argument is merely passed through to <code><a href="../reference/matrix_normalize.html">matrix_normalize()</a></code>.</p></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-7-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-7-900"></a></h3>
<ul><li>
<code><a href="../reference/sestats_to_df.html">sestats_to_df()</a></code> - simple method to convert <code>sestats</code> output from <code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> to a <code>data.frame</code> summary with number of hits for each comparison. It should be suitable for <code>kable()</code> output, with row groups by cutoff.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.6.900" id="jamses-006900">jamses 0.0.6.900<a class="anchor" aria-label="anchor" href="#jamses-006900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-6-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-6-900"></a></h3>
<ul><li>
<p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>new argument <code>sedesign</code> allows providing <code>SEDesign</code> output from <code>groups_to_design()</code>. When <code>isamples</code> is also provided, it will be used to subset the <code>sedesign</code> as appropriate, using <code>validate_design()</code>.</li>
</ul></li>
<li><p><code><a href="../reference/validate_sedesign.html">validate_sedesign()</a></code> was updated to prevent the annoying warning when comparing vectors of two different sizes.</p></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-6-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-6-900"></a></h3>
<ul><li>
<code><a href="../reference/save_sestats.html">save_sestats()</a></code> is a wrapper function intended to help export stat summary tables into Excel using <code><a href="https://rdrr.io/pkg/jamba/man/writeOpenxlsx.html" class="external-link">jamba::writeOpenxlsx()</a></code>. This function defines useful column style functions for statistical data types, such as P-values, fold changes, log fold changes, etc. It also saves all statistical contrasts, with one contrast in each worksheet. The worksheet names are limited to 31 characters, the limitation of MS Excel.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.5.900" id="jamses-005900">jamses 0.0.5.900<a class="anchor" aria-label="anchor" href="#jamses-005900"></a></h2>
<p>Bumped version, to sync the transition of slicejam functions into jamses.</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.4.900" id="jamses-004900">jamses 0.0.4.900<a class="anchor" aria-label="anchor" href="#jamses-004900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-4-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-4-900"></a></h3>
<ul><li>
<code><a href="../reference/se_collapse_by_row.html">se_collapse_by_row()</a></code> is a new function intended to collapse rows, for example peptides into protein abundance.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.3.900" id="jamses-003900">jamses 0.0.3.900<a class="anchor" aria-label="anchor" href="#jamses-003900"></a></h2>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.2.900" id="jamses-002900">jamses 0.0.2.900<a class="anchor" aria-label="anchor" href="#jamses-002900"></a></h2>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.1.900" id="jamses-001900">jamses 0.0.1.900<a class="anchor" aria-label="anchor" href="#jamses-001900"></a></h2>
<div class="section level3">
<h3 id="initial-package-0-0-1-900">initial package<a class="anchor" aria-label="anchor" href="#initial-package-0-0-1-900"></a></h3>
<p>Methods are being ported from another internal R package.</p>
</div>
<div class="section level3">
<h3 id="new-functions-0-0-1-900">New functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-1-900"></a></h3>
<ul><li><p><code><a href="../reference/matrix_normalize.html">matrix_normalize()</a></code> - wrapper function to normalize numeric matrix data</p></li>
<li><p><code><a href="../reference/se_normalize.html">se_normalize()</a></code> - function to normalize SummarizedExperiment assay data</p></li>
<li><p><code><a href="../reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> - function to convert experiment group into <code>SEDesign</code> with proper design matrix, and contrast matrix. This function was ported from <code><a href="https://rdrr.io/pkg/splicejam/man/groups2contrasts.html" class="external-link">splicejam::groups2contrasts</a></code>.</p></li>
<li><p><code><a href="../reference/se_contrast_stats.html">se_contrast_stats()</a></code> - wrapper function to perform numerous statistical contrasts, in this case calling <code>limma</code> methods with or without limma-voom.</p></li>
<li><p><code><a href="../reference/voom_jam.html">voom_jam()</a></code> - lightly extended variation of <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code> solely to enable presence of <code>NA</code> values without distortion of the model fit.</p></li>
<li><p><code><a href="../reference/handle_na_values.html">handle_na_values()</a></code> - function to manage presence of NA values using one of several appropriate strategies.</p></li>
<li><p><code><a href="../reference/run_limma_replicate.html">run_limma_replicate()</a></code> - wrapper function to perform <code>limma</code> moderated t-test steps, and to produce consistent, fully-annotated summary <code>data.frame</code>.</p></li>
<li><p><code><a href="../reference/ebayes2dfs.html">ebayes2dfs()</a></code> - internal wrapper function to convert <code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">limma::eBayes()</a></code> output to a list of <code>data.frame</code> tables.</p></li>
<li><p><code><a href="../reference/mark_stat_hits.html">mark_stat_hits()</a></code> - wrapper function used to apply one or more statistical thresholds to mark statistical hits among rows of statistical contrast results.</p></li>
<li><p><code><a href="../reference/update_function_params.html">update_function_params()</a></code> - internal function used to apply user-specific parameters to a pre-defined set of parameters stored in function formal arguments.</p></li>
<li><p><code><a href="../reference/update_list_elements.html">update_list_elements()</a></code> - internal function called by <code><a href="../reference/update_function_params.html">update_function_params()</a></code> as a generic way to update individual elements in a <code>list</code> with user-defined values.</p></li>
<li><p><code><a href="../reference/log2fold_to_fold.html">log2fold_to_fold()</a></code> and <code><a href="../reference/fold_to_log2fold.html">fold_to_log2fold()</a></code></p></li>
</ul></div>
<div class="section level3">
<h3 id="new-s4-object-class-0-0-1-900">New S4 object class<a class="anchor" aria-label="anchor" href="#new-s4-object-class-0-0-1-900"></a></h3>
<p><code>SEDesign</code>: SummarizedExperiment design object</p>
<ul><li>
<code>design</code>: numeric matrix indicating the experiment design</li>
<li>
<code>contrasts</code>: numeric matrix indicating statistical contrasts</li>
<li>
<code>samples</code>: <code>character</code> vector of sample replicates, also equivalent to <code>rownames(design)</code>
</li>
</ul><p>Some constraints:</p>
<ul><li>
<code>rownames(design)</code> contains sample replicate identifiers, intended to correspond to <code>colnames(SummarizedExperiment)</code>
</li>
<li>
<code>colnames(design)</code> contains the name of each sample group</li>
<li>
<code>rownames(contrasts)</code> contains the name of each sample group</li>
<li>Enforced constraint: <code>colnames(design) == rownames(contrasts)</code>
</li>
</ul><p>Related accessor and assignment methods:</p>
<ul><li><p><code>[</code> subset operations use <code>[samples, groups]</code> to enable re-ordering, or subsetting data by sample and/or groups. For example, one may want to remove a technical outlier sample from an analysis, or include only a subset of groups in downstream statistical contrasts.</p></li>
<li>
<p><code><a href="../reference/validate_sedesign.html">validate_sedesign()</a></code> contains the full logic for constraints and subset operations:</p>
<ul><li>the order of <code>samples</code> must match <code>rownames(design)</code>
</li>
<li>the order of <code>colnames(design)</code> must match <code>rownames(contrasts)</code>
</li>
<li>when a subset of <code>samples</code> is defined, <code>design</code> is updated to remove groups which are no longer represented by <code>min_reps</code> replicates. By default, a group with zero replicates is no longer a valid group</li>
<li>When a subset of groups is defined, perhaps by using a subset of samples as described above, the <code>contrasts</code> entries are evaluated to see which entries require the groups to be removed. For each contrast that requires a group being removed, the contrast is also removed. This step prevents retaining an incomplete contrast.</li>
</ul></li>
<li><p><code>samples()</code> will return the <code>object@samples</code></p></li>
<li><p><code>samples()&lt;-</code> to assign new values to <code>object@samples</code>, which will also update <code>rownames(design)</code> accordingly. Note these values are expected to match <code>colnames(SummarizedExperiment)</code>.</p></li>
<li><p><code>groups()</code> will return <code>colnames(object@design)</code>, representing all sample groups defined.</p></li>
<li><p><code>groups()&lt;-</code> will assign values to <code>colnames(object@design)</code>, which will also update <code>rownames(object@contrasts)</code> if there are statistical contrasts already defined in the <code>object</code>.</p></li>
<li><p><code>contrastnames()</code> will return <code>colnames(object@contrasts)</code> representing the name of each contrast.</p></li>
<li><p><code>contrastnames()&lt;-</code> will assign values to <code>colnames(object@contrasts)</code>, which may be useful to define a sensible label.</p></li>
<li><p><code>design()</code> will return the design matrix <code>object@design</code>.</p></li>
<li><p><code>design()&lt;-</code> will assign a new design matrix to <code>object@design</code>, then will call <code><a href="../reference/validate_sedesign.html">validate_sedesign()</a></code> to enforce the relevant constraints. For example <code>colnames(object@design)</code> must match <code>rownames(object@contrasts)</code> if contrasts are defined in the object.</p></li>
<li><p><code><a href="https://rdrr.io/r/stats/contrasts.html" class="external-link">contrasts()</a></code> will return the contrast matrix <code>object@contrasts</code>.</p></li>
<li><p><code>contrasts()&lt;-</code> will assign a new contrast matrix to <code>object@contrasts</code>, then will call <code><a href="../reference/validate_sedesign.html">validate_sedesign()</a></code> to enforce the relevant constraints. For example <code>colnames(object@design)</code> must match <code>rownames(object@contrasts)</code>.</p></li>
</ul></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

