<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>TODO for jamses • jamses</title><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="TODO for jamses"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">jamses</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.69.900</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="http://github.com/jmw86069/jamses/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>TODO for jamses</h1>
      <small class="dont-index">Source: <a href="http://github.com/jmw86069/jamses/blob/HEAD/TODO.md" class="external-link"><code>TODO.md</code></a></small>
    </div>

<div id="todo-for-jamses" class="section level1">

<div class="section level2">
<h2 id="id_28oct2024">28oct2024<a class="anchor" aria-label="anchor" href="#id_28oct2024"></a></h2>
<ul><li>Prepare <code>pkgdown</code> package docs.</li>
</ul></div>
<div class="section level2">
<h2 id="id_07oct2024">07oct2024<a class="anchor" aria-label="anchor" href="#id_07oct2024"></a></h2>
<ul><li>
<p>Update <code>README.Rmd</code></p>
<ul><li>Include visuals for <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>, and <code>sestats()</code>
</li>
<li>Include visuals for <code><a href="reference/heatmap_se.html">heatmap_se()</a></code>
</li>
<li>Include examples for <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code>
</li>
<li>Include description for <code><a href="reference/shrinkDataFrame.html">shrinkDataFrame()</a></code>
</li>
<li>Consider examples using <code><a href="https://rdrr.io/pkg/colorjam/man/col_div_xf.html" class="external-link">colorjam::col_div_xf()</a></code> within a customized <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">ComplexHeatmap::Legend()</a></code>.</li>
</ul></li>
<li>
<p><code><a href="reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>DONE. Consider option to populate the “hits” worksheet using <code>logFC</code> values instead of only using <code>c(1, 0, -1)</code>.</li>
</ul></li>
<li>
<p>Consider new helper functions</p>
<ul><li>
<code>contrast_list_by_factor()</code>: subdivides a set of contrasts by which factor is being compared</li>
<li>
<code><a href="reference/sestats_to_dfs.html">sestats_to_dfs()</a></code>: wrapper for <code><a href="reference/save_sestats.html">save_sestats()</a></code> which returns the list of <code>data.frames</code>.</li>
</ul></li>
<li>
<p>Consider adding conversion of <code>Seurat</code> to <code>SingleCellExperiment</code></p>
<ul><li>Key addition: maintain all assay matrix objects using their original names: <code>layername_assayname</code>
</li>
<li>Option to apply <code>log2(1 + x)</code> if range exceeds a threshold (e.g. 50)</li>
<li>Option to convert sparse <code>Matrix</code> objects to vanilla <code>matrix</code>.</li>
</ul></li>
<li>
<p>Add tests</p>
<ul><li>
<code><a href="reference/heatmap_se.html">heatmap_se()</a></code> basic workflows</li>
<li>Conversion of <code>Seurat</code> to <code>SingleCellExperiment</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_18sep2024">18sep2024<a class="anchor" aria-label="anchor" href="#id_18sep2024"></a></h2>
<ul><li><p>Adapt <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code> for <code>Seurat</code> and <code>SingleCellExperiment</code></p></li>
<li>
<p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>, <code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code></p>
<ul><li>
<p>Consider option to return/filter “useful information”:</p>
<ul><li>Filter for minimum number of replicates in a group to be permitted in a contrast.</li>
<li>Return number of replicates per group in a contrast.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_16sep2024">16sep2024<a class="anchor" aria-label="anchor" href="#id_16sep2024"></a></h2>
<ul><li>
<p>DONE. Adapt <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> for <code>SingleCellExperiment</code> objects.</p>
<ul><li>DONE. By proxy, it could also work for <code>Seurat</code> objects.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10sep2024">10sep2024<a class="anchor" aria-label="anchor" href="#id_10sep2024"></a></h2>
<ul><li>DONE. Fix error with <code><a href="reference/save_sestats.html">save_sestats()</a></code> when <code>rowData_colnames</code> contains columns already present, causing them to be added twice. The error is caused with <code>writeOpenxlsx()</code> by duplicate colnames.</li>
</ul></div>
<div class="section level2">
<h2 id="id_14aug2024">14aug2024<a class="anchor" aria-label="anchor" href="#id_14aug2024"></a></h2>
<ul><li>
<p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code> - Improve the method for bumping arrows.</p>
<ul><li>Currently all contrasts that share the same y-position or x-position are bumped relative to each other. The goal is to bump contrasts only when they overlap another contrast.</li>
<li>When B-A:X and D-C:X are on the same y-position, but one does not cross the other, they should not bump each other.</li>
<li>The new algorithm may need to account for x,y,intercept and bump number, so that a contrast is only bumped when it matches all three values. This way, when one contrast is bumped, it should free some space for other contrasts. It may also need to be adjusted iteratively.</li>
<li>Interesting to consider whether to begin with longest or shortest contrasts (by Eucliden distance) in order to control the symmetry.</li>
</ul></li>
<li>
<p>Consider method to combine two <code>SEStats</code> objects</p>
<ul><li>Typically expected to append contrasts across two objects with the same hit thresholds, assay_names. Not required, however.</li>
</ul></li>
<li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code>: Consider option to add item count to <code>row_title</code>.</p>
<ul><li>New argument? <code>tabulate_row_split=TRUE</code>
</li>
<li>Get the row split information, tabulate number of rows per split, define <code>row_title</code> to include the number in parentheses.</li>
<li>Label should be “(1,234 <code>row_type</code>)”.</li>
<li>Bonus points: Trim trailing “s” if there is only one row.</li>
</ul></li>
<li>
<p>DONE. <code><a href="reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li><p>DONE. Debug misalignment of sheet name with contrasts. Sigh.</p></li>
<li>
<p>DONE. Consider optional methods to shorten the sheet names</p>
<ul><li>custom function to edit the sheet name before truncating to <code>max_nchar_sheetname</code> number of characters. For example to edit things like <code>"male", "female"</code> to <code>"m", "f"</code> to save space.</li>
<li>option to abbreviate each term to use only the first letter or first N unique letters?</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_08aug2024">08aug2024<a class="anchor" aria-label="anchor" href="#id_08aug2024"></a></h2>
<ul><li>
<p>DONE. <code><a href="reference/matrix_normalize.html">matrix_normalize()</a></code> and <code><a href="reference/se_normalize.html">se_normalize()</a></code></p>
<ul><li>
<p>DONE. Consider option to define <code>reference_samples</code> during normalization. CORRECTION: Test existing option <code>controlSamples</code> to confirm it works as intended for this purpose. CONFIRMED.</p>
<ul><li>The goal is to normalize only the <code>controlSamples</code> to themselves, then normalize all other samples relative to that subset.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_24jun2024">24jun2024<a class="anchor" aria-label="anchor" href="#id_24jun2024"></a></h2>
<ul><li>
<p>Consider function to take <code>SEStats</code> and reverse/flip contrasts using a preferred set of contrasts.</p>
<ul><li>Use case is when receiving data with (groupA-groupB) but the preferred order is (groupB-groupA).</li>
<li>Function iterates <code>SEStats</code> and finds matching contrasts to reverse.</li>
<li>All hit signs and fold changes are multiplied by <code>-1</code> to flip the sign.</li>
<li>All matching contrast names are reversed.</li>
<li>All colnames which include the contrast also have the contrast flipped.</li>
</ul></li>
<li>
<p>Consider new function to take <code>list</code> of <code>data.frame</code> with statistical results and create <code>SEStats</code> object.</p>
<ul><li>It should parse each <code>data.frame</code> and create <code>"hit "</code> column if needed.</li>
<li>It should define <code>hit_list</code> for each <code>data.frame</code>
</li>
</ul></li>
<li>
<p>Consider function to manipulate dimnames in <code>SEStats</code>: contrast_names, assay_names, cutoff_names.</p>
<ul><li>Basic example is to convert <code>contrast</code> to <code>comp</code> or vice versa.</li>
<li>Could be through accessor <code>contrast_names(SEStats)</code> and <code>contrast_names(SEStats)&lt;-</code>
</li>
<li>Rename <code>assay_names</code> as needed.</li>
<li>Rename <code>cutoff_names</code>?? Perhaps not rename.</li>
</ul></li>
<li>
<p>Consider function to “validate” <code>SEStats</code> object, suggested rules:</p>
<ul><li>
<code>hit_array</code> contrast_names must match <code>hit_list</code> and <code>stats_dfs</code>.</li>
<li>
<code>hit_array</code> assay_names must match <code>hit_list</code> and <code>stats_dfs</code>.</li>
<li>
<code>hit_array</code> cutoff_names must match <code>hit_list</code> and <code>stats_dfs</code>.</li>
</ul></li>
<li>
<p>Consider function to <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> multiple <code>SEStats</code> objects together.</p>
<ul><li>It would combine <code>hit_list</code>, throw error whenever the input and new data both have: <code>assay_name</code>, <code>contrast_name</code>, and <code>cutoff_name</code>.</li>
</ul></li>
<li>
<p>Consider function to subset <code>SEStats</code> by dimensions: <code>assay_name</code>, <code>contrast_name</code>, <code>cutoff_name</code></p>
<ul><li>For example: <code>SEStats[1, 4:8, 2]</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_31may2024">31may2024<a class="anchor" aria-label="anchor" href="#id_31may2024"></a></h2>
<ul><li>
<p>Add testing for all variations of <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>create test case with <code>NA</code> values, to test <code>handle_na</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_30may2024">30may2024<a class="anchor" aria-label="anchor" href="#id_30may2024"></a></h2>
<ul><li>DONE. Debug <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> discrepancies when <code>isamples</code> is provided in different orders. All outputs should be identical regardless of input order. Bug was <code><a href="reference/handle_na_values.html">handle_na_values()</a></code> and was fixed, unclear when it was introduced.</li>
</ul></div>
<div class="section level2">
<h2 id="id_15may2024">15may2024<a class="anchor" aria-label="anchor" href="#id_15may2024"></a></h2>
<ul><li>
<p>DONE. <code><a href="reference/save_sestats.html">save_sestats()</a></code></p>
<ul><li>include a summary table with the <code>"hit"</code> column from each contrast</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_23apr2024">23apr2024<a class="anchor" aria-label="anchor" href="#id_23apr2024"></a></h2>
<ul><li>
<p>Design idea for <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>
<p>Allow and document how to create a multi-panel plot, with this plot as one panel in the output?</p>
<ul><li>The driving use case is to display contrasts alongside a Venn diagram or heatmap.</li>
</ul></li>
<li><p>Option to plot one contrast, showing the arrow (or arrows) and displaying at the top the full contrast name, optionally the abbreviated “comp”.</p></li>
<li><p>Potential to create a small “logo” for a single contrast, that could be displayed as a thumbnail in another slide.</p></li>
<li><p>Potential to create a small “logo” for a few contrasts, for example alongside a Venn diagram comparing two or three contrasts. The thumbnail could represent which contrasts are being displayed.</p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_16apr2024">16apr2024<a class="anchor" aria-label="anchor" href="#id_16apr2024"></a></h2>
<ul><li>
<p>Implement <code>testthis</code> unit tests for <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>Highest priority: test each option of <code>handle_na</code>
</li>
<li>Test <code>use_voom</code>.</li>
<li>Test <code>block</code>.</li>
<li>Test <code>normgroup</code>.</li>
<li>Test <code>block</code> and <code>use_voom</code> together, <code>voom_block_twostep=TRUE</code>, which should call correlation twice.</li>
<li>Test <code>block</code> and <code>normgroup</code> notably when one <code>normgroup</code> contains 2+ unique <code>block</code> values, and another <code>normgroup</code> has only 1 <code>block</code>.</li>
<li>Bonus points: Test using a subset of <code>isamples</code> that no longer contains a valid contrast. It should fail - though in future it could potentially return the group mean values, then stop short of performing contrasts.</li>
</ul></li>
<li>
<p>Extend <code>SEDesign</code> for <code>block</code> and <code>normgroup</code>:</p>
<ul><li><p>Pressing need to maintain <code>normgroup</code> and <code>block</code> with <code>sedesign</code>. Basic goal: when <code>sedesign</code> is subset, <code>normgroup</code> and <code>block</code> are also subset consistently.</p></li>
<li>
<p>New slot names:</p>
<ul><li><p><code>"normgroup"</code> - <code>character</code> vector to match <code>samples()</code>, and <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code> of the design matrix. Or could it be <code>data.frame</code> to permit multiple column values? Upon use each row would be concatenated to make one value per sample. When <code>sedesign</code> is subset, <code>"normgroup"</code> is also subset consistently.</p></li>
<li>
<p><code>"block"</code> - <code>data.frame</code> with one or more columns, indicating. Its primary purpose is to maintain values per <code>samples()</code>, so they can be subset and maintained consistently. It will be pushed to <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> how to deal with the actual values:</p>
<ul><li>
<code>character</code> values are considered <code>block</code> covariates for <code>limma</code>. For now, only <code>character</code> will be permitted.</li>
<li>
<code>numeric</code> values can be encoded as a scalar covariate, and would be appended as a new column in <code>design</code> - and therefore must be added as a new row in <code>contrasts</code> with empty <code>0</code> values.</li>
<li>
<code>integer</code>, <code>factor</code> values are encoded as an ordinal covariate, converted to rank integer values.</li>
</ul></li>
</ul></li>
<li><p><code>normgroup()</code>, <code>normgroup()&lt;-</code> - set/get functions for <code>sedesign@normgroup</code></p></li>
<li><p><code>block()</code>, <code>block()&lt;-</code> - set/get functions for <code>sedesign@block</code></p></li>
</ul></li>
<li>
<p><code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> to handle <code>normgroup</code>, <code>block</code></p>
<ul><li>When <code>normgroup</code> is supplied, contrasts should be limited to those within each <code>normgroup</code>. It may be accomplished by including <code>normgroup</code> as factor columns, but not included with <code>factor_order</code> so that comparisons cannot involve multiple <code>normgroup</code> values.</li>
<li>Store <code>normgroup</code> in the <code>SEDesign</code> object, see above.</li>
</ul></li>
<li>
<p><code>SEStats</code> S4 object for output from <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>
<p>Proposed slots:</p>
<ul><li><code>"hit_array"</code></li>
<li>
<code>"stats_dfs"</code> - each contrast in <code>data.frame</code> format</li>
<li>
<code>"stats_df"</code> - overall merged <code>data.frame</code>
</li>
<li>
<code>"sedesign"</code> - (with <code>"block"</code>, <code>"normgroup"</code>) - for reproducibility</li>
</ul></li>
<li>
<p>Methods:</p>
<ul><li><code>contrast_names()</code></li>
<li>
<code>sestats_to_list()</code> (analogous to <code><a href="reference/hit_array_to_list.html">hit_array_to_list()</a></code>)</li>
<li>
<code>sestats_to_im()</code> as above but returns signed incidence matrix</li>
<li>
<code>sestats_to_sedesign()</code> extracts equivalent <code>SEDesign</code> object</li>
</ul></li>
<li>
<p><code>hit_array()</code> - access to the array of statistical hits by dimensions:</p>
<ul><li><code>cutoff_name</code></li>
<li><code>contrast_name</code></li>
<li><code>assay_name</code></li>
<li>
<code>method_name</code> - Add this dimension to enable alternative methods</li>
</ul></li>
<li><p><code>hit_im()</code> - incidence <code>matrix</code> for specific dimensions in <code>hit_array</code></p></li>
<li><p><code>hit_list()</code> - <code>list</code> of stat hit direction, named by entity</p></li>
<li><p><code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> - <code>data.frame</code> suitable for RMarkdown and <code>kable()</code></p></li>
<li><p>accessors: <code>assay_names()</code>, <code>contrast_names()</code>, <code>cutoff_names()</code>, <code>method_names()</code></p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12mar2024">12mar2024<a class="anchor" aria-label="anchor" href="#id_12mar2024"></a></h2>
<ul><li>
<p>New function idea: <code>heatmap_to_xlsx()</code> or <code>heatmap_to_df()</code> to convert <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> output to <code>data.frame</code> or save with <code><a href="https://rdrr.io/pkg/jamba/man/writeOpenxlsx.html" class="external-link">jamba::writeOpenxlsx()</a></code>.</p>
<ul><li><p>Idea is to extract data from the heatmap as displayed, in order to save it to a file (e.g. Excel or tab-delimited) for later review.</p></li>
<li><p>Use techniques to extract the <code>left_annotation</code> and <code>top_annotation</code> data.</p></li>
<li><p>Annotation colors can be extracted and used to colorize corresponding cells in Excel.</p></li>
<li><p>It should mainly target features made available by <code><a href="reference/heatmap_se.html">heatmap_se()</a></code>, in other words any custom <code>HeatmapAnnotation()</code> functions would likely not be supported. Things like annotation bar charts, line plots, etc. would not be easily supported for export.</p></li>
<li>
<p>Main features:</p>
<ul><li><p>Rows and columns match the order as they appear in the heatmap.</p></li>
<li><p>Left annotations: <code>rowData_colnames</code>, <code>sestats</code>, <code>sestats_alt</code></p></li>
<li>
<p>Top annotations: <code>top_colnames</code></p>
<ul><li>As below, it is unclear how best to handle multiple header rows, it makes the output file much more difficult to use, whether it is saved as tab-delimited text, or Excel.</li>
</ul></li>
<li><p>Optional: Extract heatmap color function, apply to each cell. (This step sounds interesting, but is likely to be a bad idea. Applying a color per cell in Excel would be very time-consuming and inefficient, and would produce a much larger file.)</p></li>
<li><p>Optional <code>column_labels</code>, <code>row_labels</code></p></li>
<li><p>Row split encoded as a row annotation, using values from <code>row_title</code>.</p></li>
<li>
<p>Column split - unclear the best approach to use:</p>
<ul><li>Adding column annotations requires saving multiple header rows, which makes the exported file substantially more difficult to use.</li>
<li>One option is to use column split as a prefix, e.g. “Group 1 - colname1”, “Group 1 - colname2”, “Group 2 - colname3”, etc.</li>
</ul></li>
<li><p>Option to supply the <code>se</code> object, and append additional data via <code>rowData_colnames</code>.</p></li>
</ul></li>
<li>
<p>Required: Ability to save <code>HeatmapList</code> objects, already drawn using <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">ComplexHeatmap::draw()</a></code>.</p>
<ul><li>Bonus points: recognize when multiple heatmaps are present, e.g. from <code>ComplexHeatmap::draw(hm1 + hm2 + hm3)</code>.</li>
<li>It needs to iterate each heatmap in the <code>HeatmapList</code>, potentially also obtaining <code>left_annotation</code> and <code>top_annotation</code> from each <code>Heatmap</code>. Ugh. Each heatmap could have different <code>top_annotations</code> which means the <code>top_annotation</code> would need to have a column with row names for each heatmap’s unique <code>top_annotation</code>.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_07mar2024">07mar2024<a class="anchor" aria-label="anchor" href="#id_07mar2024"></a></h2>
<ul><li>
<p><code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code></p>
<ul><li>Consider adding argument <code>normgroup</code> to restrict contrasts within each <code>normgroup</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code>: Consider some way to indicate <code>controlSamples</code>.</p>
<ul><li>Note that sometimes <code>controlSamples</code> are not displayed on the heatmap, this might be an exception where it is not shown. The rule might be: if not all <code>controlSamples</code> are in <code>isamples</code> then do not indicate <code>controlSamples</code>.</li>
<li>One option is to render an asterisk just above each column? <code>"*"</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_06mar2024">06mar2024<a class="anchor" aria-label="anchor" href="#id_06mar2024"></a></h2>
<ul><li>
<p><code><a href="reference/se_normalize.html">se_normalize()</a></code></p>
<ul><li>
<p>DONE. Consider allowing user-specified <code>assay_name</code> to store the resulting data? Some mechanism to allow customizing the output <code>assay_name</code>. New arguments:</p>
<ul><li><p>DONE. <code>output_method_prefix</code> which by default uses each value in <code>method</code> to formulate each output <code>assay_name</code></p></li>
<li>
<p>DONE. <code>output_assay_names</code></p>
<ul><li>specify completely user-defined output <code>assay_name</code> values, overriding <code>method</code> and <code>output_method_prefix</code> when defining the output <code>assay_name</code>.</li>
<li>
<code>character</code> vector, must be equal to the number of normalizations. Note that each normalization is applied to each <code>assay_names</code> in order: <code>method1_assay1</code>, <code>method1_assay2</code>, <code>method2_assay1</code>, <code>method2_assay2</code>.</li>
</ul></li>
</ul></li>
<li>
<p>DONE. Consider using <code>mcols(assays(se))</code> to store annotation regarding the normalization:</p>
<ul><li>DONE. store values: <code>"normalization_method"</code>, <code>"source_assay_name"</code>, <code>"params"</code> (params may not be easy to use, since params is a <code>list</code> whose elements may include <code>character</code> vectors, <code>numeric</code> values with many decimal places, etc.</li>
<li>DENIED. Consider comparing new <code>mcols()</code> entries to existing <code>mcols()</code> entries to decide whether to overwrite the existing entry, or to create a new <code>assays()</code> entry with versioned name and different params. Decision: Do <strong>not</strong> make this change, push this back onto users to define <code>output_method_prefix</code> or <code>output_assay_names</code>, with the idea that the user should specifically request creating a new <code>assay_name</code>.</li>
</ul></li>
<li><p>DENIED. If <code>method="jammanorm"</code> is called with different params, should it create a new <code>assay_name</code> entry associated with those params? <strong>Decision: No.</strong></p></li>
</ul></li>
<li>
<p><code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>Consider permitting a subset of <code>se</code> columns being passed, to allow labeling only a subset of columns per operation. Ignore columns which are not provided in the <code>se</code> object.</li>
<li>DONE. Consider option to hide the group line, when the group label is also empty or whitespace.</li>
</ul></li>
<li>
<p><code>heatmap_row_group_labels()</code></p>
<ul><li>New function completely analogous to <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code>.</li>
<li>Default <code>rot</code> text rotation is 180 degrees (sideways).</li>
<li>Consider adding something like <code>hm_title_buffer</code> to adjust the buffer to the left of <code>left_annotation</code>, <code>HeatmapBody</code>?</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_26feb2024">26feb2024<a class="anchor" aria-label="anchor" href="#id_26feb2024"></a></h2>
<ul><li>
<p><code><a href="reference/se_normalize.html">se_normalize()</a></code></p>
<ul><li>
<p>DONE. new methods from <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code>:</p>
<ul><li>“TMM”: trimmed mean of M-values</li>
<li>“TMMwsP”: modified TMM with singleton pairing, removes zeros</li>
<li>“RLE”: relative log expression</li>
<li>“upperquartile”: uses upper 75% (by default) or user-defined threshold</li>
</ul></li>
<li>
<p>DENIED. Consider composite <code>assay_name</code> with param values?</p>
<ul><li>for example running the same normalization with two sets of <code>params</code>
</li>
</ul></li>
</ul></li>
<li>
<p>DENIED. consider renaming <code>"limma_batch_effect"</code></p>
<ul><li>something shorter, and without underscores</li>
<li>
<code>"lba"</code>, <code>"limmabatchadjust"</code>
</li>
<li>(Instead user can customize the resulting assay_name.)</li>
</ul></li>
<li>
<p>DELAYED. Consider something like <code>list_normalization_methods()</code></p>
<ul><li>purpose is to list available normalizations as a programmatic reference</li>
<li>returns <code>data.frame</code>: abbreviation, full_name, description</li>
<li>the <code>abbreviation</code> ultimately becomes the new <code>assay_name</code>
</li>
<li>Currently <code>jamba::jargs(se_normalize, "method")</code> will print available options, except without descriptive information. <code><a href="reference/se_normalize.html">? se_normalize</a></code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_23feb2024">23feb2024<a class="anchor" aria-label="anchor" href="#id_23feb2024"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>
<p>When an entry in <code>sample_color_list</code> contains a color function, the breaks are taken from <code>attr(x, "breaks")</code>, and it uses the same values as labels.</p>
<ul><li>Consider recognizing optional attribute with “labels” to use: <code>attr(x, "labels")</code>. It must be the same length as <code>"breaks"</code>.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_25jan2024">25jan2024<a class="anchor" aria-label="anchor" href="#id_25jan2024"></a></h2>
<ul><li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>when using blocking factor (argument <code>block</code>) the group mean values slightly differ from what could be calculated manually, even when applying limma batch adjustment.</li>
<li>Excellent overview of experiment design and contrast models in R: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7873980/" class="external-link uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7873980/</a> doi: 10.12688/f1000research.27893.1</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10jan2024">10jan2024<a class="anchor" aria-label="anchor" href="#id_10jan2024"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>DONE. Consider using <code>"legend_title"</code> instead of <code>"name"</code> as the heatmap legend title, so that <code>"name"</code> can be defined with a specific value. The driving need is when adding two heatmaps together, if they both have the same <code>"name"</code> the name appears twice when calling <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/list_components.html" class="external-link">ComplexHeatmap::list_components()</a></code>. Instead, we should allow <code>name</code> to be defined uniquely for each heatmap, even while both heatmaps share the same legend title.</li>
<li>Consider convenient method and default to add separation between heatmap and annotations, e.g. <code>ht_opt(ROW_ANNO_PADDING=grid::unit(4, "mm"))</code> and <code>ht_opt(COLUMN_ANNO_PADDING=grid::unit(4, "mm"))</code>. It should set the value, then revert to previous value afterward.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_08jan2024">08jan2024<a class="anchor" aria-label="anchor" href="#id_08jan2024"></a></h2>
<ul><li>DONE: debug <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> with <code>handle_na="full1"</code>
</li>
</ul></div>
<div class="section level2">
<h2 id="id_11dec2023">11dec2023<a class="anchor" aria-label="anchor" href="#id_11dec2023"></a></h2>
<ul><li>
<p><code><a href="reference/contrasts_to_venn_setlists.html">contrasts_to_venn_setlists()</a></code></p>
<ul><li>DONE. Assign list names based upon the contents of each Venn setlist, instead of the default names that are not user-friendly.</li>
<li>Consider “two-way” Venn subsets that include: two-way contrast, corresponding one-way contrasts.</li>
</ul></li>
<li>
<p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>When <code>contrast_depths=2</code> and supplying <code>sestats</code> it labels the one-way and two-way contrasts, but should only label one-way contrasts. The workaround is to alter <code>sestats$hit_array</code> to include only the contrasts being displayed.</li>
</ul></li>
<li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>consider some method to “group” the <code>sestats</code> contrasts, for example sub-grouping them by which factor(s) are being compared. Use <code><a href="reference/contrasts_to_factors.html">contrasts_to_factors()</a></code> to generate a table, then split by which factors have comparisons (with delimiter <code>"-"</code>).</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_04dec2023">04dec2023<a class="anchor" aria-label="anchor" href="#id_04dec2023"></a></h2>
<ul><li>
<p><code><a href="reference/contrast2comp.html">contrast2comp()</a></code> and <code><a href="reference/contrast2comp.html">comp2contrast()</a></code></p>
<ul><li>
<p>consider method that can convert contrasts inside stat colnames:</p>
<ul><li>
<code>"logFC factorA_factorB-factorC-factorB"</code> to <code>"logFC factorA-factorD:factorB"</code>
</li>
<li>
<code>"hit mgm5 adjp0.05 fc1 PNU_Control_Nano147car1-Veh_Control_Nano147car1"</code> to <code>"hit mgm5 adjp0.05 fc1 PNU-Veh:Control:Nano147car1"</code>
</li>
<li>It therefore needs to detect the location of the contrast or comp, convert that substring, then replace with the new value.</li>
</ul></li>
</ul></li>
<li>
<p>migrate <code><a href="https://rdrr.io/pkg/platjam/man/design2colors.html" class="external-link">platjam::design2colors()</a></code></p>
<ul><li>
<p>Consider adding SE-specific method</p>
<ul><li>assign colors to <code>rowData()</code> and <code>colData()</code> in one step</li>
<li>bonus points for sharing color assignments for shared terms</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_09nov2023">09nov2023<a class="anchor" aria-label="anchor" href="#id_09nov2023"></a></h2>
<ul><li><p>add documentation for generic methods, <code><a href="https://rdrr.io/r/stats/contrasts.html" class="external-link">contrasts()</a></code>, <code>contrast_names()</code>, <code>groups()</code>, <code>samples()</code>, etc.</p></li>
<li>
<p>consider DESeq2 support</p>
<ul><li>
<p>Research the equivalent design model, contrasts model for equivalent comparisons in DESeq2 as used with limma-voom. Often people seem to include only groups relevant to each contrast in the DESeq2 workflow, unclear if this is recommended guidance.</p>
<ul><li>Guidance is similar to limma-voom, include each sample group as relevant to the overall comparisons, using <code>normgroup</code> to separate subsets that are expected to differ substantially by sample type, or where each subset may represent different variabilities.</li>
<li>Therefore, this option would initially follow the limma-voom default with user-defined, optional use of <code>normgroup</code> which can define independent subsets of sample groups.</li>
<li>There should probably be an option to perform each contrast in its own independent subset, which would not be equivalent to <code>normgroup</code>, for example “untreated”, “treated1”, “treated2” would imply three one-way contrasts, and each contrast would be performed in its own unique group-to-group analysis.</li>
</ul></li>
<li>
<p>Include options specific to DESeq2:</p>
<ul><li>lfcShrink method</li>
<li>normalization method (or none)</li>
<li>optional outlier removal (e.g. Cook’s distance method)</li>
</ul></li>
<li><p>Decide how to handle stats column headers which differ from limma, so they are easily used in downstream methods: <code>"P.Value"</code>, <code>"adj.P.Val"</code>, <code>"logFC"</code>, etc.</p></li>
<li><p>Consider returning dispersion data sufficient to produce the dispersion QC plot to review. (Consider doing similar for <code>voom()</code> internals.)</p></li>
<li><p>Consider including “method” as a dimension in <code>hit_array</code>, so that limma-voom, DESeq2 could both be used and directly compared. Similarly the <code>posthoc_method="DEqMS"</code> may be included so that its effects can be compared to limma without the custom posthoc adjustment.</p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_18oct2023">18oct2023<a class="anchor" aria-label="anchor" href="#id_18oct2023"></a></h2>
<ul><li>
<p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>Option to filter by <code>max_depth</code> (to show only oneway contrasts) or <code>contrast_depths</code> (to show only oneway, or only twoway contrasts).</li>
<li>When <code>group_buffer</code> is adjusted, the replicate <code>"n=3"</code> labels are not shifted, so they can appear outside the square.</li>
<li>Consider option to adjust aspect ratio.</li>
<li>Consider option to adjust the offset between contrasts.</li>
</ul></li>
<li>
<p><code><a href="reference/filter_contrast_names.html">filter_contrast_names()</a></code></p>
<ul><li>Consider option to enable sequential comparisons, for example: Time1-Time0, Time2-Time1, Time3-Time2. This option may be more useful when enabled only for certain factors. Not as useful for something like “Treatment” where each treatment usually does not have a sequential order.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_06oct2023">06oct2023<a class="anchor" aria-label="anchor" href="#id_06oct2023"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Add optional argument to apply heatmap title to <code>column_title</code>, so that it appears above the heatmap automatically. <strong>It will override showing <code>column_split</code> titles.</strong> So it may be best practice to use <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code>
</li>
<li>Consider argument to add whitespace padding to the heatmap title, to make it more convenient to use <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_04oct2023">04oct2023<a class="anchor" aria-label="anchor" href="#id_04oct2023"></a></h2>
<ul><li>Done. fix bug in <code><a href="reference/contrasts_to_venn_setlists.html">contrasts_to_venn_setlists()</a></code> resulting in some sets with &gt;4 entries.</li>
<li>Consider integrating <code>sestats</code> with <code><a href="reference/contrasts_to_venn_setlists.html">contrasts_to_venn_setlists()</a></code> so that the output is actually a <code>list</code> suitable for <code><a href="https://rdrr.io/pkg/venndir/man/venndir.html" class="external-link">venndir::venndir()</a></code>.</li>
</ul></div>
<div class="section level2">
<h2 id="id_02oct2023">02oct2023<a class="anchor" aria-label="anchor" href="#id_02oct2023"></a></h2>
<ul><li>
<p><code>contrast_names()&lt;-</code></p>
<ul><li>validate the supplied <code>value</code> and print error message when incompatible.</li>
</ul></li>
<li>
<p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>consider easy filter to hide two-way contrasts</li>
<li>consider scaling the arrow width (decreased) when there is a large number of contrasts to display, especially when large number are “bumped” per factor level.</li>
<li>consider making <code>arrow_ex</code>,<code>head_ex</code> arguments to <code>make_block_arrow()</code> visible in <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>. Perhaps change <code>twoway_lwd</code> to <code>twoway_cex</code>, then applying some global <code>cex</code> to all contrast arrows and connectors.</li>
<li>Done. Fix bug with custom contrasts, two-way contrast color <code>NA</code> throws error.</li>
</ul></li>
<li>
<p><code><a href="reference/validate_sedesign.html">validate_sedesign()</a></code></p>
<ul><li>argument <code>contrasts</code> is not properly subsetting contrasts by name, however <code>contrast_names()&lt;-</code> appears to be working fine. Same mechanism.</li>
</ul></li>
<li>
<p>Done. <code><a href="reference/filter_contrast_names.html">filter_contrast_names()</a></code></p>
<ul><li>new function: take long list of “all versus all” contrasts and subset for specific control factor levels.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_20sep2023">20sep2023<a class="anchor" aria-label="anchor" href="#id_20sep2023"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>consider indicating <code>controlSamples</code> when a subset of samples are used, as opposed to relying upon <code>control_label</code>. Especially useful when a subset of potential samples are used, or when there are multiple normgroups displayed, each with their own <code>controlSamples</code>.</li>
<li>consider option to apply <code>hm_title</code> to <code>column_title</code> which may help when adding two heatmaps together, they would have the column_title defined within each heatmap itself. Downside: no column titles defined by <code>column_split</code>.</li>
</ul></li>
<li>
<p><code>SEDesign</code>: consider associating blocking factor to one or more contrasts</p>
<ul><li>Unclear how: blocking factor is per-sample annotation</li>
<li>
<code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> does not indicate blocking factor - this is the real need, so output represents the comparison.</li>
</ul></li>
<li>
<p><code>SEDesign</code>: consider associating factor name with each contrast</p>
<ul><li>
<p>Background: Each contrast compares one or more factors. It would be useful to “know” the factors being compared during automated analysis.</p>
<ul><li>Contrasts could be subset by factor(s) being compared</li>
<li>Center data using other factors in the design: when showing Treatment contrasts, could center by Time or Genotype</li>
<li>Venn diagrams make the most sense when comparing within factor: Treatment contrasts to compare across time points or genotypes. Gene knockout comparisons across treatments (Vehicle, DEX, Etoposide)</li>
</ul></li>
<li>
<p>Proposed changes to <code>SEDesign</code></p>
<ul><li><p>new slot: <code>"contrast_factors"</code>, <code>list</code> named by <code>colnames(contrasts)</code> values contain zero or more <code>colnames(factors)</code></p></li>
<li>
<p>new slot: <code>"factors"</code>, <code>data.frame</code></p>
<ul><li>
<code>colnames(factors)</code> are experimental factors (Treatment, Time)</li>
<li>values are factor levels inferred during <code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code>
</li>
<li>rows are samples, <code>rownames(factors) == rownames(design)</code>
</li>
</ul></li>
<li>
<p>subsetting SEDesign</p>
<ul><li>by sample would also subset <code>factors</code>
</li>
<li>by group would also subset <code>contrasts</code> and <code>contrast_factors</code>
</li>
</ul></li>
</ul></li>
</ul></li>
<li>
<p><code>SEDesign</code>: consider supporting <code>~Treatment + Time</code> style design matrix</p>
<ul><li><p>Current design and contrast matrices use <code>~ 0 + Treatment + Time</code>, so that contrasts indicate distinct experiment groups.</p></li>
<li>
<p><code>~Treatment + Time</code> format changes:</p>
<ul><li>
<code>design</code> includes <code>"(Intercept)"</code> then factor levels</li>
<li>instead of groupA-groupB contrasts, it encodes coefficients using factor levels.</li>
</ul></li>
<li><p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code> needs to handle this format differently, visualizing factor comparisons differently as well.</p></li>
<li><p>Two-factor designs might compare one factor, which implies using the other factor as a covariate term, equivalent to modeling one factor with the second factor as a blocking factor.</p></li>
<li><p>Unclear how to represent a specific interaction term if encoded as <code>~Treatment + Time + Treatment:Time</code></p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_31aug2023">31aug2023<a class="anchor" aria-label="anchor" href="#id_31aug2023"></a></h2>
<ul><li>
<p>Method to <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code> or <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code> SummarizedExperiment objects.</p>
<ul><li><p>Main goal is to automate the process of aligning samples</p></li>
<li><p>Keep or remove metadata columns that do not match</p></li>
<li>
<p>When metadata does not match:</p>
<ol style="list-style-type: decimal"><li>Remove the metadata column altogether (easiest option), or</li>
<li>Keep metadata, but “combine” values by comma-delimiting (easiest), or by some numerical operation (complex to design R function interface).</li>
</ol></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_25aug2023">25aug2023<a class="anchor" aria-label="anchor" href="#id_25aug2023"></a></h2>
<ul><li>
<p>Create <code>SEStats</code> object as more formal S4 object output <code>se_contrast_stats</code>()</p>
<ul><li>
<p>slotNames</p>
<ul><li><p><code>stats_dfs</code>: each <code>data.frame</code> from each <code>contrast_name</code>, <code>assay_name</code></p></li>
<li>
<p><code>hit_array</code>: N-dimensional array of stat hits with direction:</p>
<ul><li><code>assay_name</code></li>
<li><code>contrast_name</code></li>
<li><code>cutoff_name</code></li>
<li>
<code>method_name</code>? (to compare limmavoom, limma, DESeq2, edgeR?) adding this dimension could be fairly invisible</li>
</ul></li>
<li><p><code>stats_df</code>: <strong>omit</strong> - separate function to merge <code>data.frame</code></p></li>
<li><p><code>metadata</code>: method, parameters, etc.</p></li>
</ul></li>
<li>
<p>methods</p>
<ul><li>
<code>hit_list()</code>: calls <code><a href="reference/hit_array_to_list.html">hit_array_to_list()</a></code>
</li>
<li>
<code>to_df()</code>: calls <code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> to create table summary of counts</li>
<li>
<code>hits()</code>: converts <code>hit_list()</code> into incidence matrix?</li>
<li>
<code>contrast_names()</code>: extracts <code>contrast_name</code> vector</li>
<li>
<code>assay_names()</code>: extracts <code>assay_name</code> vector</li>
<li>
<code>reapply_cutoffs()</code>: (new) re-calculate <code>hit_array</code>, by iterating <code>stats_dfs</code> and applying stat cutoffs.</li>
<li>
<code>rbind_sestats()</code>: combines multiple <code>SEStats</code> objects</li>
</ul></li>
<li>
<p>Other related todo:</p>
<ul><li>
<p>consider backward compatibility?</p>
<ul><li>
<code>SEStats_to_list()</code> to convert <code>SEStats</code> to previous <code>list</code> format</li>
<li>
<code>list_to_SEStats()</code> to convert previous format to new <code>SEStats</code>
</li>
</ul></li>
<li><p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code> should accept <code>SEStats</code> input</p></li>
<li><p><code><a href="reference/hit_array_to_list.html">hit_array_to_list()</a></code> should accept <code>SEStats</code> input</p></li>
<li><p><code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> should accept <code>SEStats</code> input</p></li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>use new <code>SEStats</code> object input</li>
<li>consider expanding <code>sestats</code> to show separate <code>cutoff_name</code> and <code>method_name</code> entries as distinct stripes in the incidence matrix, rather than including hits across <code>cutoff_name</code> entries together. The labels could become quite long.</li>
</ul></li>
<li><p>Expand <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> to call corresponding DESeq2 methods.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_22aug2023">22aug2023<a class="anchor" aria-label="anchor" href="#id_22aug2023"></a></h2>
<ul><li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>
<p>For very large data volume, the method seems to take more memory than absolutely necessary and could be trimmed:</p>
<ul><li>Remove <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code> for <code>stats_dfs</code> and <code>stats_df</code>. Row identifiers are roughly 50% the size of each <code>data.frame</code>, so they should not be stored in a column and as rownames. The rownames are less “safe” to R manipulation, so values will be retained in a specific column (first column). Other functions that utilize <code>data.frame</code> objects must use column values and not rely upon rownames, in theory should already be true.</li>
<li>Option to omit <code>stats_df</code>, roughly 25% overall object size. Matter of fact, this object could be replaced by a function that converted <code>stats_dfs</code> into <code>stats_df</code> dynamically.</li>
<li>option to omit <code>stats_dfs</code>, roughly 75% overall object size, but used to create volcano plots, and to review specific results.</li>
</ul></li>
<li>
<p>When using <code>block</code> the process becomes substantially slower with large data.</p>
<ul><li>
<p>Description of the scenario, and supporting evidence:</p>
<ul><li>The slow step occurs during <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code> when supplied with <code>block</code> and when <code>correlation=NULL</code>, so it is calculated inside <code>lmFit()</code>.</li>
<li>The actual rate-limiting step is <code><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">limma::duplicateCorrelation()</a></code> which is run to determine <code>correlation</code>.</li>
<li>In one test case with 220k rows, 30 columns, this step took 7 minutes.</li>
<li>A subset with a random 10k rows took 15 seconds, and estimated the same summary correlation value.</li>
<li>When provided the pre-calculated <code>correlation</code> the <code>lmFit()</code> using 220k rows took 1.5 seconds.</li>
</ul></li>
<li>
<p>Potential workarounds:</p>
<ol style="list-style-type: decimal"><li>When <code>block</code> is defined, and <code>correlation</code> is not supplied, calculate <code>correlation</code> using a subset of up to N rows (e.g. 10000). The correlation could even be calculated 10 times using random subsets of 1000 rows, then take the average. The max rows could be a new argument <code>max_correlation_rows=10000</code> to make this process explicit, and customizable.</li>
<li>The process above could be “pushed back” to any functions calling <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> to avoid having an invisible difference between using this function, and using <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code>
</li>
</ol></li>
</ul></li>
<li><p>Need to store arguments such as <code>correlation</code>, and <code>block</code> alongside the returned <code>sestats</code> object.</p></li>
</ul></li>
<li><p>Debug why loading <code>jamses</code> causes the warning to the effect: <code>"design() has already been defined"</code>.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_14aug2023">14aug2023<a class="anchor" aria-label="anchor" href="#id_14aug2023"></a></h2>
<ul><li>
<p><code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>Option to size block arrows by the relative number of hits?</li>
<li>Option to define block arrow sizes as a vector, one per arrow, which probably means one per contrast name.</li>
</ul></li>
<li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Consider option to cluster rows by stats hit matrix data. Unclear how, but would be useful to sub-group hits.</li>
</ul></li>
<li>
<p><code><a href="reference/se_normalize.html">se_normalize()</a></code>: use <code>mcols(assays(se))</code>:</p>
<ul><li>
<p>Driving use case:</p>
<ul><li>somewhere to store parameters used during analysis.</li>
<li>
<code>metadata(assays(se)[[1]])</code> cannot be used, since metadata on the matrix itself is easily lost upon any sort of manipulation, subsetting of the matrix.</li>
<li>secondary use case: ability to filter for normalized data</li>
</ul></li>
<li><p>upon creating a new entry in <code>assays(se)</code> it should also populate <code>mcols(assays(se))</code> with columns of annotation. These are optional annotations, but could be convenient for including things like arguments to the methods used.</p></li>
<li><p>Consider updating <code>platjam</code> data import methods similarly.</p></li>
<li>
<p>Suggested annotation names:</p>
<ul><li>
<code>"method"</code>: name of the normalization method applied</li>
<li>
<code>"params"</code>: <code>list</code> of parameters for the given method, obtained from the argument with <code>params[[method]]</code>
</li>
<li>
<code>"parent_assay_name"</code>: may not be practical. The <code>assay_name</code> can be edited, which would invalidate the relationship.</li>
<li>
<code>preferred</code>: optional flag to indicate the preferred <code>assay_name</code> for downstream analysis?</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/se_normalize.html">se_normalize()</a></code>: change default <code>output_sep="_"</code> to <code>output_sep="."</code>?</p>
<ul><li>
<p>benefit is that method names (containing underscores) are more easily distinguished when concatenated:</p>
<ul><li>
<code>"jammanorm.limma_batch_adjust.totalIntensity"</code> versus</li>
<li><code>"jammanorm_limma_batch_adjust_totalIntensity"</code></li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_27jul2023">27jul2023<a class="anchor" aria-label="anchor" href="#id_27jul2023"></a></h2>
<ul><li>
<p>Remaining TODO for <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>:</p>
<ul><li><p>Fix error when any of axis1,axis2,axis3,axis4 are assigned empty values.</p></li>
<li>
<p>Options for rendering two-way contrasts:</p>
<ul><li>
<p>PARTIAL. Consider option to “loop” two-way contrasts around the one-way contrast, similar to what happens when two contrasts are too close to each other.</p>
<ul><li>Loops are implemented but did not meet the favor of artistic review.</li>
<li>For contrasts with very small gap between them, loop seems the best.</li>
<li>For contrasts with room for an S-shaped swoop, that looks nicer.</li>
</ul></li>
<li><p>Option to adjust the midpoint of S-shaped “swoop”, so the middle of the swoop would not be the diagonal midpoint between the end of contrast 1 and start of contrast 2. It could be shifted closer to contrast 1 or contrast 2. It could help when trying to avoid label overlaps.</p></li>
</ul></li>
<li><p>Call <code><a href="reference/sedesign_to_factors.html">sedesign_to_factors()</a></code> instead of calculating internally.</p></li>
<li><p>Consider refactoring the drawing order so that contrasts are drawn, then all labels are (potentially) drawn atop the graphics so they are more consistently visible. (Could be done if porting to grid graphics.)</p></li>
<li><p>Accept contrast_names as input, instead of requiring an <code>sedesign</code> object.</p></li>
<li><p>Consider using grid graphics with with the <code>units="snpc"</code> pattern to maintain fixed aspect ratio:</p></li>
</ul><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pushViewport</span><span class="op">(</span></span>
<span><span class="fu">viewport</span><span class="op">(</span></span>
<span>   x<span class="op">=</span><span class="fl">0.5</span>, y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>   width<span class="op">=</span><span class="fu">unit</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">xlim</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">ylim</span><span class="op">)</span><span class="op">)</span>, <span class="st">"snpc"</span><span class="op">)</span>,</span>
<span>   height<span class="op">=</span><span class="fu">unit</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">ylim</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">xlim</span><span class="op">)</span><span class="op">)</span>, <span class="st">"snpc"</span><span class="op">)</span>,</span>
<span>   xscale<span class="op">=</span><span class="va">xlim</span>,</span>
<span>   yscale<span class="op">=</span><span class="va">ylim</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ul><li>Consider drawing boxes which reflect the <code>groupedAxis()</code> grouped regions along each axis, which may involve slightly nested boxes. It may be useful to shade the boxes light grey.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12jul2023">12jul2023<a class="anchor" aria-label="anchor" href="#id_12jul2023"></a></h2>
<ul><li>
<p>enhance <code>sedesign</code></p>
<ul><li>Consider adding <code>factor_labels</code> to have design labels for each factor in the group label.</li>
</ul></li>
<li>
<p>DONE (initial implementation): new function <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code></p>
<ul><li>Simplified version of <code>plotComparisonTable()</code> from past work.</li>
<li>Factor levels should be “known”.</li>
<li>Layout should follow logic similar to (or directly extending) <code><a href="https://rdrr.io/pkg/vcd/man/mosaic.html" class="external-link">vcd::mosaic()</a></code> which defines factors on x-axis, y-axis, then subdivides each axis in order with sub-factors.</li>
<li>Layout should define every observed combination of factor levels.</li>
<li>Then block arrows (or regular arrows) can be used to indicate each pairwise contrast.</li>
<li>Block arrows are “bumped” when they would overlap another block arrow.</li>
<li>Bonus points: Block arrow labels can include the contrast name, or comp name (shortened version of the contrast), and optionally the number of statistical hits.</li>
<li>Bonus points: Indicate the number of replicates per group.</li>
<li>Two-way contrasts would indicate both one-way contrasts, with a “ribbon” connecting the end of the first contrast to the beginning of the next contrast. Something like a beta-sheet in protein structure graphics.</li>
<li>Most contrasts should only be vertical or horizontal, therefore non-standard contrasts would be angled, indicating that they are comparing more than one factor at a time.</li>
</ul></li>
<li>
<p>Completed TODO for <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>:</p>
<ul><li><p>DONE. Option to indicate “n=8” number of replicates per group.</p></li>
<li><p>DONE. Provide or add custom label; option to hide labels.</p></li>
<li><p>DONE. Show subset of contrasts based upon contrast name.</p></li>
<li><p>DONE. Consider option to “flip” two-way contrasts: (A-B)-(C-D) is equivalent to (A-C)-(B-D) and can be shown either way.</p></li>
<li><p>DONE. Decide how to assign colors to contrasts. For now, colors are assigned to one-way contrasts, then two-way contrasts inherit two colors as a gradient.</p></li>
<li><p>DONE. Adjust axis labels to be adjacent to the square boxes used per group. This step used argument <code>pos</code> which places the axis at a fixed coordinate inside the plot.</p></li>
<li><p>DONE. Allow custom position for contrast labels, scaled from 0 to 1 indicating the position along each contrast (from-to) to place the text label. This step also applies text justification, which helps prevent text from spilling beyond the relevant end of the arrow.</p></li>
<li>
<p>DONE. Add argument <code>sestats</code> to print the hits per contrast.</p>
<ul><li>DONE. Define how to handle multiple <code>assay_names</code>, <code>cutoff_names</code>. It passes arguments to <code><a href="reference/hit_array_to_list.html">hit_array_to_list()</a></code>. Anything more complicated requires the user to pass argument with custom labels.</li>
</ul></li>
<li>
<p>DONE. Sort contrasts:</p>
<ul><li>by axis display order; currently axes are ordered in the order that factor levels appear in the contrasts.</li>
<li>then by “distance” along each axis, so longer contrasts are “bumped” consistently.</li>
</ul></li>
<li><p>DONE. Consider making hit direction optional when <code>sestats</code> is used.</p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_27jun2023">27jun2023<a class="anchor" aria-label="anchor" href="#id_27jun2023"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Consider automating color assignment when <code>top_colnames</code> or <code>rowData_colnames</code> have no colors assigned in <code>sample_color_list</code>. It may involve calling <code><a href="https://rdrr.io/pkg/platjam/man/design2colors.html" class="external-link">platjam::design2colors()</a></code> with empty argument for <code>group_colnames</code>, which may involve moving that function into <code>colorjam</code>.</li>
</ul></li>
<li>
<p>DONE: <code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>DONE: when <code>correlation=TRUE</code> the label with number of rows should indicate number of columns? See below, both dimensions are indicated. Or number of rows used to calculate correlation of N columns?</li>
<li>DONE: Consider new argument <code>column_type="samples"</code> and default heatmap title also indicates the number of columns (samples).</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_05jun2023">05jun2023<a class="anchor" aria-label="anchor" href="#id_05jun2023"></a></h2>
<ul><li>
<p>DONE: <code><a href="reference/se_collapse_by_column.html">se_collapse_by_column()</a></code></p>
<ul><li>DONE: consider changing default <code>noise_floor_value=NA</code> to <code>noise_floor_value=0</code>.</li>
<li>DONE: consider new argument <code>useMedian=FALSE</code> so that the default behavior is to take the mean and not the median value per group.</li>
<li>DONE: change to use <code>jamba::call_fn_ellipse()</code> so that calls to row group functions will only pass arguments accepted by that function.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_19may2023">19may2023<a class="anchor" aria-label="anchor" href="#id_19may2023"></a></h2>
<ul><li>
<p>DONE: add <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code></p>
<ul><li>Custom function to augment <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> for a specific scenario, but the scenario is used often enough to warrant making the function available here.</li>
<li>The function draws group labels above a rendered heatmap, with underline drawn by default indicating the heatmap columns under each label. It draws multiple levels of group labels when defined.</li>
<li>It uses <code>grid</code> coordinates <em>after rendering</em> to determine where to position labels, which requires drawing the heatmap first.</li>
<li>Also the group labels cannot be included in the <code>Heatmap</code> object, which is not ideal, but is also a known limitation.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_05may2023">05may2023<a class="anchor" aria-label="anchor" href="#id_05may2023"></a></h2>
<ul><li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li><p>DONE: add optional <code>rowData()</code> colnames to stat <code>data.frame</code> output, for example adding <code>rowData_colnames=c("SYMBOL", "GENENAME")</code> would keep gene symbol and gene name alongside microarray probe IDs.</p></li>
<li><p>in future, optionally run DESeq2 equivalent steps to limma/limma-voom, by replacing the <code><a href="reference/run_limma_replicate.html">run_limma_replicate()</a></code> step with optional function to wrapper DESeq2 steps.</p></li>
</ul></li>
<li><p>tests for <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> with various <code>handle_na</code> values, and with/without <code>rowData_colnames</code>.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_05apr2023">05apr2023<a class="anchor" aria-label="anchor" href="#id_05apr2023"></a></h2>
<ul><li><p>DONE: use <code>testthat</code> unit testing</p></li>
<li>
<p><code>sedesign</code> object</p>
<ul><li>DONE: add method <code>contrastNames()</code> (or <code>contrast_names()</code>)</li>
<li>consider adding <code>comps()</code> as shortcut for <code><a href="reference/contrast2comp.html">contrast2comp()</a></code>
</li>
</ul></li>
<li>
<p><code><a href="reference/contrast2comp.html">contrast2comp()</a></code></p>
<ul><li>optimize performance, it is surprisingly slow, but functional</li>
<li>consider embedding the factor order into the output for two-way contrasts, to ensure the output exactly matches input even when alternate outputs are mathematically equivalent.</li>
</ul></li>
<li>
<p>add <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>, <code>plot_contrasts()</code></p>
<ul><li>display design with similar layout orientation to <code><a href="https://rdrr.io/pkg/vcd/man/mosaic.html" class="external-link">vcd::mosaic()</a></code>
</li>
<li>indicate contrasts with block arrows drawn between groups</li>
<li>optionally label groups by number of replicates</li>
<li>optionally label contrasts by number of statistical hits</li>
</ul></li>
<li>
<p>DONE: add <code><a href="reference/contrast_names_to_sedesign.html">contrast_names_to_sedesign()</a></code></p>
<ul><li>convenience function to produce <code>sedesign</code> from only <code>contrastNames</code>
</li>
<li>mostly helpful for demo purposes, development of <code><a href="reference/plot_sedesign.html">plot_sedesign()</a></code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_28mar2023">28mar2023<a class="anchor" aria-label="anchor" href="#id_28mar2023"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>Add optional title above <code>sestats</code> incidence matrix display, which could address need to display which <code>assay_name</code>, and <code>cutoff_name</code> was used.</li>
<li>Might be useful to have additional option to provide a label to display atop the <code>sestats</code> incidence matrix.</li>
<li>Probably needs other options such as font sizing, label rotation, etc. Oi.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12mar2023">12mar2023<a class="anchor" aria-label="anchor" href="#id_12mar2023"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code> improvements</p>
<ul><li>DONE: Some method to hide <code>column_title</code> labels. Use <code>column_title=" "</code>.</li>
<li>Consider <code>hmgrouplabel</code> function <code><a href="reference/heatmap_column_group_labels.html">heatmap_column_group_labels()</a></code> which enables cleaner column labels when the heatmap is split by one or more variables in <code>colData(se)</code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_22nov2022">22nov2022<a class="anchor" aria-label="anchor" href="#id_22nov2022"></a></h2>
<ul><li>
<p>Add concept of <code>"normgroup"</code> to <code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code></p>
<ul><li>Implied: <code>"normgroup"</code> should also be added to <code>sestats</code>.</li>
</ul></li>
<li><p>Add per-gene logic and functions developed for the DM-JDM manuscript.</p></li>
<li>
<p>Bonus points: optionally print commands as they are being run, for example:</p>
<ul><li>general idea is that <code>se_contrasts()</code> is a wrapper around limma, so it should be able to print the equivalent calls to <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code> for example, so a user can observe the progression of analysis steps.</li>
<li>command to convert SE to expression matrix / and ExpressionSet used by limma</li>
<li>command to run optional <code><a href="reference/voom_jam.html">voom_jam()</a></code>; then the vanilla <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code>
</li>
<li>command to run <code><a href="reference/run_limma_replicate.html">run_limma_replicate()</a></code>
</li>
<li>command to run <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code>, <code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">limma::eBayes()</a></code>
</li>
</ul></li>
<li><p>COMPLETE: previous update to <code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> introduced a regression (error) for input that generates two-way contrasts.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_20oct2022">20oct2022<a class="anchor" aria-label="anchor" href="#id_20oct2022"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li>when <code>centerby_colnames=FALSE</code> no data centering is performed, the color legend should match the range of data, and hide negative values if there are no negative values.</li>
<li>allow custom <code>col</code> color function in form of <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">circlize::colorRamp2()</a></code>
</li>
<li>deprecate <code>sample_color_list</code> into <code>color_list</code>; or consider adding <code>row_color_list</code> so row colors can contain the same colnames with different color assignments compared with <code>sample_color_list</code>. Hmmm.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_13oct2022">13oct2022<a class="anchor" aria-label="anchor" href="#id_13oct2022"></a></h2>
<ul><li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code></p>
<ul><li>COMPLETE: argument <code>normgroup</code> to enforce independent statistical analyses within each unique normgroup.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_21sep2022">21sep2022<a class="anchor" aria-label="anchor" href="#id_21sep2022"></a></h2>
<ul><li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> to include additional gene annotation data</p>
<ul><li>similarly, the first column currently hardcoded <code>"probes"</code>, should use the appropriate column name, or allow it to be defined by argument.</li>
<li>rowname is currently the only identifier included in results.</li>
<li>implementation ideas: it could either use input <code>rowData(se)</code>, in the limma model fit, or annotation can be added while each <code>stats_df</code> stat <code>data.frame</code> is created.</li>
</ul></li>
<li>
<p><code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code></p>
<ul><li>(COMPLETE) Mechanism to supply specific contrast names to be used in place of auto-generated contrasts.</li>
<li>(COMPLETE) Allow <code>SummarizedExperiment</code> input, with <code>group_colnames</code> used to define sample groups.</li>
</ul></li>
<li>
<p><code>sestats</code> object output from <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code>:</p>
<ul><li><p>proper slot names: stats_df, stats_dfs, hit_array, hit_list, sedesign</p></li>
<li><p>proper print method that calls <code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> by default.</p></li>
<li>
<p>functions:</p>
<ul><li>
<code>hit_array(sestats)</code> with arguments assay_name, cutoff, contrasts</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> bug:</p>
<ul><li>apparently the colnames do not match the dimnames, <code>"cutoff"</code> is displayed for <code>"assay_name"</code> values. Probably something was reversed during recursive <code>list</code> navigation.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12sep2022">12sep2022<a class="anchor" aria-label="anchor" href="#id_12sep2022"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code> does not have option to customize arguments <code>row_names_gp</code> nor <code>column_names_gp</code>, which could be used to colorize, highlight, boldface, individual labels in the heatmap.</p>
<ul><li>Currently <code>row_names_gp</code> is defined internally, in order to define <code>fontsize</code> based upon the number of rows and columns in the heatmap. The <code>fontsize</code> could be applied to user-defined <code>row_names_gp</code> or <code>column_names_gp</code>, however sometimes the rows and columns are defined dynamically - making it difficult to sync a vector of <code>grid::gpar(col=c("red", "black"))</code> values to the exact rownames.</li>
<li>To accommodate dynamic rows, it might need another method such as named vectors for known <code><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">grid::gpar()</a></code> attributes: <code>col</code>, <code>fontsize</code>, <code>fill</code>, <code>alpha</code>, <code>fontfamily</code>, <code>fontface</code>, <code>cex</code>, <code>font</code> (<code>font</code> is an alias for <code>fontface</code>, should be passed as <code>fontface</code>).</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_03aug2022">03aug2022<a class="anchor" aria-label="anchor" href="#id_03aug2022"></a></h2>
<ul><li>
<p><code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> implement normalization groups with these rules:</p>
<ul><li>No one-way contrast, direct comparison, should be permitted which compares to different normalization groups.</li>
<li>Two-way contrasts are permitted across normalization groups, only those that involve one-way comparisons within a normalization group.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_18jul2022">18jul2022<a class="anchor" aria-label="anchor" href="#id_18jul2022"></a></h2>
<ul><li><p>Some form of power calculation that leverages the same methods used by <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code>. It could leverage the <code>ssizeRNA</code> package for RNA-seq data.</p></li>
<li><p><code><a href="reference/contrast2comp.html">contrast2comp()</a></code> is somehow very slow for even 10 to 20 contrast names.</p></li>
<li>
<p><code><a href="reference/contrast2comp.html">contrast2comp()</a></code> adjustments to tolerate having label prefix.</p>
<ul><li>for example <code>"fold (A_c-B_c)(A_d-B_d)"</code> could be recognized as <code>"label contrast"</code> and return <code>"fold A-B:c-d"</code>.</li>
<li>main assumption is no spaces in contrast name</li>
<li>may be useful to specify whether to look for label prefix or suffix, in the event a label contains a hyphen or dash <code>"-"</code> character.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_11jul2022">11jul2022<a class="anchor" aria-label="anchor" href="#id_11jul2022"></a></h2>
<ul><li>
<code><a href="reference/heatmap_se.html">heatmap_se()</a></code> font size customizations sufficient for manuscript prep.</li>
</ul></div>
<div class="section level2">
<h2 id="id_27jun2022">27jun2022<a class="anchor" aria-label="anchor" href="#id_27jun2022"></a></h2>
<ul><li>COMPLETE: <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> when not supplied <code>rows</code> nor <code>sestats</code> results in an error.</li>
<li>COMPLETE: <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> should have option not to center data.</li>
<li>COMPLETE: <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> should allow custom incidence matrix through <code>sestats</code> and <code>alt_sestats</code>, to avoid having to provide <code>sestats</code> or <code>hit_array</code>.</li>
</ul></div>
<div class="section level2">
<h2 id="id_22jun2022">22jun2022<a class="anchor" aria-label="anchor" href="#id_22jun2022"></a></h2>
<ul><li>COMPLETE: <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> - ability to “drill-down” into row clusters</li>
</ul></div>
<div class="section level2">
<h2 id="id_21jun2022">21jun2022<a class="anchor" aria-label="anchor" href="#id_21jun2022"></a></h2>
<ul><li>COMPLETE: Bug in <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> when <code>sestats</code> is supplied but <code>se</code> does not contain all rows present in <code>sestats</code> hit array.</li>
<li>COMPLETE: <code><a href="reference/heatmap_se.html">heatmap_se()</a></code> needs control over the annotation name fontsize.</li>
</ul></div>
<div class="section level2">
<h2 id="id_16jun2022">16jun2022<a class="anchor" aria-label="anchor" href="#id_16jun2022"></a></h2>
<ul><li>
<p><code><a href="reference/heatmap_se.html">heatmap_se()</a></code></p>
<ul><li><p>COMPLETE: argument <code>isamples</code> should be useful to define samples to display in the heatmap. However, it would be useful to perform centering before subsetting by samples, in order to produce more useful graphs with paired data. For example, center data by each patient at time zero, then display the other timepoints (since the patient time zero would always be exactly zero, it only contributes a blank stripe to the heatmap).</p></li>
<li>
<p>COMPLETE: <code>top_annotation</code> and <code>left_annotation</code> shows color key for all colors, not just the those which are displayed in the heatmap annotation.</p>
<ul><li>The potential downside is that it might affect results when two heatmaps are added together using <code>+</code> or <code>%v%</code> - in those cases equivalent color keys are sometimes merged together. Unsure how it would be handled when they are only partially identical.</li>
<li>UPDATE: It actually merges color keys by name, and displays unique colors for each name. That’s so awesome.</li>
</ul></li>
<li><p>COMPLETE: create new function to choose interesting annotation colnames, with logic that removes columns with 1:1 cardinality compared to other chosen columns. Columns that only repeat the information are no longer interesting.</p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_24may2022">24may2022<a class="anchor" aria-label="anchor" href="#id_24may2022"></a></h2>
<ul><li>
<code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> consider making a wider output format intended for <code>kable</code>, with relevant columns grouped by <code>assay_name</code>. This way the output includes one row per contrast.</li>
<li>COMPLETE: <code><a href="reference/sestats_to_df.html">sestats_to_df()</a></code> should report blank cell whenever NA values are present, to convey that no cutoff was applied for that scenario, rather than implying the cutoff was applied and there were no hits.</li>
</ul></div>
<div class="section level2">
<h2 id="id_26apr2022">26apr2022<a class="anchor" aria-label="anchor" href="#id_26apr2022"></a></h2>
<ul><li>
<p><code><a href="reference/se_detected_rows.html">se_detected_rows()</a></code> was added, however</p>
<ul><li>in future would be nice to apply constraints based upon contrasts. (In hindsight, I’m not sure what use case I had in mind!)</li>
</ul></li>
<li>
<p>COMPLETE: Migrate the <code>volcano_plot()</code> from <code>slicejam::volcano_plot()</code></p>
<ul><li>currently also draws block arrows in the plot margins</li>
<li>Migrated into <code>jamma</code> package, alongside MA-plots.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_18apr2022">18apr2022<a class="anchor" aria-label="anchor" href="#id_18apr2022"></a></h2>
<ul><li>
<p>New class <code>sestats</code> to replace output from <code>se_contrast_stats</code></p>
<ul><li>
<p>slotNames:</p>
<ul><li>hit_array</li>
<li>hit_list</li>
<li>stat_dfs</li>
<li>stat_df</li>
</ul></li>
<li>
<p>subset: [signal, contrast, cutoff]</p>
<ul><li>intended to help filter for specific results</li>
<li>when supplying <code>sestats[matrix(ncol=3)]</code> it will subset for each element in the <code>hit_array</code>
</li>
</ul></li>
<li>
<p>accessors:</p>
<ul><li>
<code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code>, <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> prints the <code>data.frame</code> summary of hit counts</li>
<li>
<code>hits()</code> will return <code>list</code> of <code>list</code>
</li>
<li>
<code>hit_array()</code> will return the full <code>sestats@hit_array</code>
</li>
<li>
<code>hit_im()</code> will return an incidence matrix of hits</li>
</ul></li>
<li>
<p>converters</p>
<ul><li>
<code><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list()</a></code> to convert to previous <code>list</code> format</li>
<li>
<code>list2sestats()</code> to convert from previous <code>list</code> format to <code>sestats</code>
</li>
</ul></li>
</ul></li>
<li>
<p>COMPLETE: Some method to rename two-way contrasts to save character space.</p>
<ul><li>
<p>Notes:</p>
<ul><li>
<p>Contrasts should be easily distinguished, since <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> encodes the contrast into column headers, for example</p>
<ul><li><code>"logFC CellA_Treated-CellA_Control"</code></li>
<li><code>"adj.P.Value CellA_Treated-CellA_Control"</code></li>
<li><code>"P.value CellA_Treated-CellA_Control"</code></li>
<li><code>"hit mgm5 adjP0.01 fc1.5 CellA_Treated-CellA_Control"</code></li>
<li>Currently “easy” to find the contrast by taking the last non-whitespace string from the column header.</li>
<li>Contrasts should have no whitespace, use <code>":"</code> delimiter?</li>
</ul></li>
<li>
<p>Renamed contrasts ideally do not use parentheses, since the goal is to reduce characters.</p>
<ul><li>Technically <code>(CellA_Treated-CellA_Control)-(CellB_Treated-CellB_Control)</code>
</li>
<li>Equivalent: <code>CellA_Treated-CellA_Control:CellB_Treated-CellB_Control</code>
</li>
<li>Also equivalent: <code>Treated-Control:CellA-CellB</code>
</li>
</ul></li>
</ul></li>
<li>
<p>Two-way contrast:</p>
<ul><li>
<code>(CellA_Treated-CellA_Control)-(CellB_Treated-CellB_Control)</code> (59 characters, 27 per contrast)</li>
</ul></li>
<li>
<p>Alternative two-way syntax:</p>
<ul><li><code>Treated-Control x CellA-CellB</code></li>
<li><code>Treated-Control:CellA-CellB</code></li>
</ul></li>
<li>
<p>Two-way with one unchanging factor:</p>
<ul><li>
<code>(CellA_Treated_WT-CellA_Control_WT)-(CellB_Treated_WT-CellB_Control_WT)</code> (71 characters, 33 per contrast)</li>
</ul></li>
<li>
<p>Alternative two-way with extra factor:</p>
<ul><li>
<code>CellA-CellB:Treated-Control:WT</code> (32 characters)</li>
<li><code>Treated-Control CellA-CellB WT</code></li>
</ul></li>
<li>
<p>One-way contrast with extra factor:</p>
<ul><li>
<code>CellA_Treated_WT-CellA_Control_WT</code> (33 characters)</li>
</ul></li>
<li>
<p>Alternative one-way with extra factor:</p>
<ul><li>
<code>Treated-Control:CellA:WT</code> (24 characters)</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_14apr2022">14apr2022<a class="anchor" aria-label="anchor" href="#id_14apr2022"></a></h2>
<p>Simpler methods for common visualizations</p>
<ul><li><p>COMPLETE: expression heatmap - using <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code></p></li>
<li>
<p>Venn diagram of hits - using <code><a href="https://rdrr.io/pkg/venndir/man/venndir.html" class="external-link">venndir::venndir()</a></code></p>
<ul><li>design idea: use second order contrasts to determine which contrasts are “compatible” to be used in the same Venn diagrams.</li>
<li>contrast in cell type A; compared to same contrast in cell type B</li>
<li>one treatment-control in cell type A; another treatment-control cell type A</li>
<li>implied: it would not generally choose treatmentA-control cellA; treatmentB-control cellB</li>
<li>it could compare two-way contrasts using third-order contrast logic</li>
</ul></li>
<li><p>COMPLETE: volcano plots - migrate function from slicejam into <code>jamma</code> package.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_13apr2022">13apr2022<a class="anchor" aria-label="anchor" href="#id_13apr2022"></a></h2>
<ul><li>
<p>COMPLETE: <code><a href="reference/save_sestats.html">save_sestats()</a></code> is super slow, for 6 worksheets, ~15 columns, 25k rows, took about 5 minutes. Should be much faster.</p>
<ul><li>Likely imposes changes to <code><a href="https://rdrr.io/pkg/jamba/man/writeOpenxlsx.html" class="external-link">jamba::writeOpenxlsx()</a></code>
</li>
<li>Is conditional formatting is the slow step? If so, skip it.</li>
<li>Is categorical formatting is the slow step? If so, skip it.</li>
<li>Changes were made in <code><a href="https://rdrr.io/pkg/jamba/man/writeOpenxlsx.html" class="external-link">jamba::writeOpenxlsx()</a></code> to operate on an open Workbook without saving, passing the Workbook to each internal step also without saving. Each worksheet is added to the Workbook, and it is only saved at the end. Saved about 20x time especially for large multi-sheet Workbooks.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_11apr2022">11apr2022<a class="anchor" aria-label="anchor" href="#id_11apr2022"></a></h2>
<ul><li>
<p>COMPLETE: New function: <code><a href="reference/save_sestats.html">save_sestats()</a></code> or something similar.</p>
<ul><li>Saves statistical results to Excel <code>.xlsx</code> file.</li>
<li>One contrast per worksheet.</li>
<li>Uses <code><a href="https://rdrr.io/pkg/jamba/man/writeOpenxlsx.html" class="external-link">jamba::writeOpenxlsx()</a></code> and defines each column type.</li>
<li>Also calls <code><a href="https://rdrr.io/pkg/jamba/man/set_xlsx_colwidths.html" class="external-link">jamba::set_xlsx_colwidths()</a></code> to set proper column widths.</li>
<li>TODO: Optionally saves the superwide <code>data.frame</code> stat table output.</li>
</ul></li>
<li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> enhancements</p>
<ul><li><p>Consider object type <code>sestats</code> so it can have proper <code>print.sestats()</code> and <code>summary.sestats()</code> generic functions.</p></li>
<li><p>When using blocking factor with voom, follow guidance to apply voom/contrasts using the two-step process.</p></li>
<li>
<p>Some method to enforce <code>normgroup</code></p>
<ul><li>each group should be analyzed independently</li>
<li>results should be collated together into one <code>hit_array</code>
</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/groups_to_sedesign.html">groups_to_sedesign()</a></code> needs some method to define first-order contrasts distinct from second-order contrasts.</p>
<ul><li>
<p>The second-order contrasts appear to use the wrong <code>factor_order</code>, e.g.</p>
<ul><li>first order contrasts: <code>"A_Treat-A_Veh"</code> and <code>"B_Treat-B_Veh"</code>
</li>
<li>second order contrast observed:<code>"(B_Treat-A_Treat)-(B_Veh-A_Veh)"</code>
</li>
<li>second order contrast intended:<code>"(B_Treat-B_Veh)-(A_Treat-A_Veh)"</code>
</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code>, <code><a href="reference/se_normalize.html">se_normalize()</a></code>, <code><a href="reference/matrix_normalize.html">matrix_normalize()</a></code> need optional argument <code>normgroup</code></p>
<ul><li><p>COMPLETE: <code><a href="reference/se_normalize.html">se_normalize()</a></code> and <code><a href="reference/matrix_normalize.html">matrix_normalize()</a></code></p></li>
<li><p>each subset of samples is processed independently within each <code>normgroup</code>.</p></li>
<li><p>intended to allow batch processing while also keeping distinct subsets of samples independent of one another.</p></li>
<li>
<p>basic workflow:</p>
<ul><li>Apply method to entire data, optional, to create the full output matrix.</li>
<li>Iterate each <code>normgroup</code> subset, apply method, populate each result into the full output matrix.</li>
<li>In the case of batch adjustment, if <code>normgroup</code> subset does not work for batch adjustment (only one batch is represented) then copy data as-is into the output matrix.</li>
</ul></li>
</ul></li>
<li><p>Consider some mechanism to hold blocking factor within <code>sedesign</code> object.</p></li>
</ul></div>
<div class="section level2">
<h2 id="contrast-design-plot">contrast design plot<a class="anchor" aria-label="anchor" href="#contrast-design-plot"></a></h2>
<ul><li><p>x,y axes represent design factors, with factor levels at each position</p></li>
<li><p>draw block arrow for each pairwise contrast, one group to another group</p></li>
<li><p>Bonus points: use hit_array to plot hit counts inside each block arrow</p></li>
<li><p>each block in the grid represents a group</p></li>
<li><p>probably use fixed aspect=1 to ensure blocks are square</p></li>
<li><p>if there is one factor, use it as x-axis, and use y-axis with no label</p></li>
<li><p>if there are more than two factors, use them as a pseudo-z-axis with slight offset</p></li>
<li>
<p>when two block arrows are co0linear, fan them out from the center</p>
<ul><li>co-linear: identical slope (m) and intercept (b)</li>
<li>distance between parallel lines: abs(b1 - b2) / sqrt(1 + m^2)</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="stats-result-object-type">stats result object type?<a class="anchor" aria-label="anchor" href="#stats-result-object-type"></a></h2>
<ul><li>
<p>Goal would be to store output from <code><a href="reference/se_contrast_stats.html">se_contrast_stats()</a></code> in a proper S4 object:</p>
<ul><li>stats_df</li>
<li>stats_dfs</li>
<li>hit_array</li>
<li>SEDesign (samples, design, contrasts used)</li>
</ul></li>
<li><p>Method to access stats data.frame by contrast</p></li>
<li><p>Method to access statistical hits</p></li>
<li><p>Method to review SEDesign used for the analysis</p></li>
<li><p>Method to convert/aggregate probe-level hits to gene level, if relevant</p></li>
<li><p>Method to use alongside SE data to visualize data</p></li>
<li><p>Method to re-apply stat thresholds using stored results</p></li>
</ul></div>
<div class="section level2">
<h2 id="add-capability-to-se_contrast_stats-to-call-deseq2">Add capability to <code>se_contrast_stats()</code> to call DESeq2<a class="anchor" aria-label="anchor" href="#add-capability-to-se_contrast_stats-to-call-deseq2"></a></h2>
<ul><li>wrap contrasts,design into DESeq2-compatible workflow</li>
<li>store stat table output from DESeq2 similar to current method</li>
<li>store relevant DESeq2-specific output</li>
</ul></div>
<div class="section level2">
<h2 id="add-capability-to-se_contrast_stats-to-test-isoformsexons">Add capability to <code>se_contrast_stats()</code> to test isoforms/exons<a class="anchor" aria-label="anchor" href="#add-capability-to-se_contrast_stats-to-test-isoformsexons"></a></h2>
<ul><li>wraps <code><a href="https://rdrr.io/pkg/limma/man/diffSplice.html" class="external-link">limma::diffSplice</a></code> or <code>DEXSeq</code> differential effects per gene</li>
</ul></div>
<div class="section level2">
<h2 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a></h2>
<ul><li>Vignette idea: Practical guide to creating gene expression heatmaps using ComplexHeatmap.</li>
</ul></div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

